---
title: "Comparing mouse and Bornstein _et al._ mouse thymus"
output: html_notebook
---

# Introduction

Comparing our ageing mouse thymus data with mouse TEC from Bornstein _et al_. Nature 2018. I need to do this on just the WT TEC cells.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(igraph)
library(SingleCellExperiment)
library(scran)
library(scater)
library(Matrix)
library(umap)
library(irlba)
library(scales)
library(viridis)
library(ggplot2)
library(biomaRt)
library(ggthemes)
library(batchelor)
library(FNN)
library(pheatmap)
library(reshape2)
library(ggalluvial)
source("~/Dropbox/R_sessions/SingleCellFunctions/single_cell_functions.R")
source("~/Dropbox/R_sessions/GGMike/theme_mike.R")
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.sce <- readRDS(file="~/Dropbox/AgeingExperiment/Frozen/Bornstein_SCE.RDS")
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.meta <- read.table("~/Dropbox/Thymus/Nature_paper/GSE103967_metadata.txt.gz",
                         sep="\t", header=TRUE, stringsAsFactors=FALSE)
```

Normalisation is already done. There are no cell type labels attached to these data :(

```{r}
sink(file="/dev/null")
gc()
sink(file=NULL)
```

Now HVGs first.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.hvg <- find_hvg(bornstein.sce, p.threshold=1e-3, return.ranks=TRUE)
table(bornstein.hvg$HVG)
```

~3000 HVGs across all Bornstein mouse thymus cells - I need to subset to just the WT ones to make a valid comparison with our mouse TEC.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
wildtype.cells <- bornstein.meta$well[bornstein.meta$Experiment_ID %in% c("Thymic_Epithel_6d", "Thymic_Epithel_newMarkers")]
wildtype.sce <- bornstein.sce[, intersect(colnames(bornstein.sce), wildtype.cells)]

saveRDS(wildtype.sce, file="~/Dropbox/AgeingExperiment/Human_Thymus/WT_Bornstein-mouse_SCE.RDS")
wildtype.hvg <- find_hvg(wildtype.sce, p.threshold=1e-3, return.ranks=TRUE)
table(wildtype.hvg$HVG)
```

There ~3000 cells from the WT thymi.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
wildtype.pca <- prcomp_irlba(as.matrix(t(logcounts(wildtype.sce[wildtype.hvg$HVG, ]))), n=50)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
wildtype.umap <- as.data.frame(umap(as.matrix(wildtype.pca$x[, c(1:20)]),
                                     n_neighbors=11,
                                     init='spectral',
                                     metric='cosine',
                                     n_components=2,
                                     min_dist=0.1,
                                     method='naive')$layout)

colnames(wildtype.umap) <- paste0("UMAP", c(1:2))
wildtype.umap$CellID <- colnames(wildtype.sce)

wildtype.meta.umap <- merge(bornstein.meta, wildtype.umap, by.x='well', by.y='CellID')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(wildtype.meta.umap, aes(x=UMAP1, y=UMAP2, colour=Experiment_ID)) +
  geom_point(size=0.1) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(override.aes=list(size=3), ncol=1))
```

These are all of the WT cells from the stroma and specifically epithelia - I currently don't have any cell type assignments for them.


```{r, echo=FALSE, warning=FALSE, message=FALSE}
goi <- c("Ccl21a", "Krt5", "Aire", "Cd52", "Dclk1", "Avil", "Syngr1", "Chga", "Cd177", "Psmb11", "Prss16", "Krt80", "Spink5", "Lifr",
         "Ly6d", "Ly6e", "Ly6a", "Fgfr2", "Fgfr3", "Cdk1", "Pdpn", "Plekhb1", "Itgb2", "Dll4", "Cdh2", "Gas1", "Igfbp4", "Mgp", "Apoe", "Gsn", "Anxa1")
goi.exprs <- as.data.frame(as.matrix(t(logcounts(wildtype.sce[intersect(rownames(wildtype.sce), goi), ]))))
goi.exprs$CellID <- colnames(wildtype.sce)

wildtype.goi.merge <- merge(wildtype.meta.umap, goi.exprs, by.x='well', by.y='CellID')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(wildtype.goi.merge, aes(x=UMAP1, y=UMAP2, colour=Ccl21a)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```

Based on _Ccl21a_ it looks like the intertypical TEC/mTEC(I) are quite prominent

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(wildtype.goi.merge, aes(x=UMAP1, y=UMAP2, colour=Lifr)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```

_Lifr_ is also a bit sporadic but overlaps mostly with the _Ccl21a_.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(wildtype.goi.merge, aes(x=UMAP1, y=UMAP2, colour=Ly6a)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```

This is _Ly6a_, again note the overlap - I think the Intertypical TEC are a subset of the mTEC(I) population.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(wildtype.goi.merge, aes(x=UMAP1, y=UMAP2, colour=Pdpn)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```

_Pdpn_ is most evident in the fibroblasts, but some sporadic expression through the mTEC(I) compartment as well.

Let's try something a bit different, namely use our mouse thymus HVGs to define the Teichmann mouse TEC manifold. 

```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.hvgs <- read.table("~/Dropbox/AgeingExperiment/Thymus_HVG.tsv",
                         sep="\t", header=TRUE, stringsAsFactors=FALSE)
table(mouse.hvgs$HVG)
```

There are just shy of 5000 mouse HVGs. I'll use these to define the manifold in the Teichmann mouse cells.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
biomaRt.connection <- useMart("ensembl", "mmusculus_gene_ensembl")

gene.df <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "chromosome_name"),
                 mart=biomaRt.connection)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.hvg.bornstein <- intersect(rownames(wildtype.sce), gene.df$external_gene_name[gene.df$ensembl_gene_id %in% mouse.hvgs$gene_id[mouse.hvgs$HVG]])

mouse.homolog.pca <- prcomp_irlba(as.matrix(t(logcounts(wildtype.sce[mouse.hvg.bornstein, ]))), n=50)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.homolog.umap <- as.data.frame(umap(as.matrix(mouse.homolog.pca$x[, c(1:20)]),
                                     n_neighbors=7,
                                     init='spectral',
                                     metric='cosine',
                                     n_components=2,
                                     min_dist=0.75,
                                     method='naive')$layout)

colnames(mouse.homolog.umap) <- paste0("Homolog.UMAP", c(1:2))
mouse.homolog.umap$CellID <- colnames(wildtype.sce)

wildtype.homolog.umap <- merge(wildtype.goi.merge, mouse.homolog.umap, by.x='well', by.y='CellID')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(wildtype.homolog.umap, aes(x=Homolog.UMAP1, y=Homolog.UMAP2, colour=Experiment_ID)) +
  geom_point(size=0.1) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(override.aes=list(size=3), ncol=1))
```

Our mouse HVGs still capture a lot of the variance in the Bornstein data as well. OK, I need to do this concretely and map the Bornstein and our 
mouse data together.

## Combining bornstein and our mouse data

```{r, echo=FALSE, warning=FALSE, message=FALSE}
thymus.meta <- read.table("~/Dropbox/AgeingExperiment/Data/ThymusMarker_tSNE_PCA_meta.tsv",
                          sep="\t", h=TRUE, stringsAsFactors=FALSE)

# create a variable that represents the ages as a single continuous variable
thymus.meta$AgeInt <- as.numeric(gsub(thymus.meta$Age, pattern="wk", replacement=""))
thymus.meta$AgeFactor <- ordered(thymus.meta$Age,
                                 levels=c("1wk", "4wk", "16wk", "32wk", "52wk"))
thymus.meta$TFIDF.Cluster <- factor(thymus.meta$TFIDF.Cluster,
                                    levels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10"),
                                    labels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10"))

thymus.meta <- thymus.meta[!thymus.meta$TFIDF.Cluster %in% c(4), ]
rownames(thymus.meta) <- thymus.meta$Sample
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.sce <- readRDS("~/Dropbox/AgeingExperiment/Data/Ageing_Mouse_SCE.RDS")
```

First I'll combine the Bornstein and mouse SCE objects using a set of commonly expressed 1:1 homologous genes.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.genes <- rownames(mouse.sce)
bornstein.genes <- rownames(wildtype.sce)

match.gene.df <- gene.df[gene.df$external_gene_name %in% bornstein.genes & gene.df$ensembl_gene_id %in% mouse.genes, ]
match.gene.df <- match.gene.df[!duplicated(match.gene.df$external_gene_name), ]
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# map the bornstein genes to ensembl IDs
rownames(match.gene.df) <- match.gene.df$external_gene_name
bornstein.homolog.sce <- wildtype.sce[unique(match.gene.df$external_gene_name), ]
rownames(bornstein.homolog.sce) <- match.gene.df[rownames(bornstein.homolog.sce), ]$ensembl_gene_id

rownames(match.gene.df) <- match.gene.df$ensembl_gene_id
mouse.homolog.sce <- mouse.sce[unique(match.gene.df$ensembl_gene_id), ]
rownames(match.gene.df) <- match.gene.df$external_gene_name
# colData(mouse.homolog.sce) <- NULL
# rowData(mouse.homolog.sce) <- NULL
# 
# # need to set this for each dimension
# reducedDim(mouse.homolog.sce) <- NULL
# reducedDim(mouse.homolog.sce) <- NULL
# reducedDim(mouse.homolog.sce) <- NULL

combine.homolog.sce <- do.call(cbind, list("mouse"=mouse.homolog.sce, "bornstein"=bornstein.homolog.sce))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# make a combined meta-data
bornstein.combine.meta <- wildtype.goi.merge[, c("well", "Experiment_ID")]

rownames(thymus.meta) <- thymus.meta$Sample
mouse.combine.meta <- thymus.meta[colnames(mouse.homolog.sce), c("Sample", "TFIDF.Cluster")]
mouse.combine.meta$cell.types <- "Unknown"
mouse.combine.meta$cell.types <- "Unknown"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "2"] <- "Intertypical TEC"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "9"] <- "Perinatal cTEC"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "3"] <- "Mature cTEC"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "7"] <- "Mature mTEC"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "1"] <- "Post-Aire mTEC"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "5"] <- "Tuft-like mTEC"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "6"] <- "Proliferating TEC"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "8"] <- "nTEC"
mouse.combine.meta$cell.types[mouse.combine.meta$TFIDF.Cluster == "10"] <- "sTEC"

colnames(bornstein.combine.meta) <- c("Sample", "CellType")
colnames(mouse.combine.meta) <- c("Sample", "Cluster", "CellType")
bornstein.combine.meta$Species <- "Mouse.Bornstein"
mouse.combine.meta$Species <- "Mouse"

mouse.bornstein.meta <- do.call(rbind.data.frame, list("bornstein"=bornstein.combine.meta,
                                                       "mouse"=mouse.combine.meta[, c("Sample", "CellType", "Species")]))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
rownames(match.gene.df) <- match.gene.df$external_gene_name
mouse.hvg.ensembl <- match.gene.df[mouse.hvg.bornstein, ]$ensembl_gene_id
mouse.hvg.ensembl <- mouse.hvg.ensembl[!is.na(mouse.hvg.ensembl)]
combine.homolog.pca <- prcomp_irlba(as.matrix(t(logcounts(combine.homolog.sce[mouse.hvg.ensembl, ]))), n=50)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
combine.homolog.umap <- as.data.frame(umap(as.matrix(combine.homolog.pca$x[, c(1:20)]),
                                           n_neighbors=7,
                                           init='spectral',
                                           metric='cosine',
                                           n_components=2,
                                           min_dist=0.75,
                                           method='naive')$layout)

colnames(combine.homolog.umap) <- paste0("Combine.UMAP", c(1:2))
combine.homolog.umap$Sample <- colnames(combine.homolog.sce)
combine.umap.merge <- merge(mouse.bornstein.meta, combine.homolog.umap, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(combine.umap.merge, aes(x=Combine.UMAP1, y=Combine.UMAP2, colour=Species)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_colorblind() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))
```

OK - so firstly there is a clear batch effect that needs to be dealt with.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=8.95}
ggplot(combine.umap.merge, aes(x=Combine.UMAP1, y=Combine.UMAP2, colour=CellType)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=2, title.position="top", override.aes=list(size=3)))
```

I don't have celltype information on the Bornstein data - so this is the experiment ID instead.

### Bornstein - mouse batch correction

The first step is a cosine normalisation of the gene expression before PCA.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
cosine.bornstein.mouse <- cosineNorm(logcounts(combine.homolog.sce), subset.row=mouse.hvg.ensembl)
cosine.pca <- prcomp_irlba(t(cosine.bornstein.mouse), n=50)
```

I'll then batch correct the bornstein data together.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.cells <- combine.umap.merge$Sample[combine.umap.merge$Species %in% c("Mouse.Bornstein")]
mouse.cells <- combine.umap.merge$Sample[combine.umap.merge$Species %in% c("Mouse")]

bornstein.cosine <- cosine.bornstein.mouse[, bornstein.cells]
mouse.cosine <- cosine.bornstein.mouse[, mouse.cells]
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.cosine.meta <- combine.umap.merge[combine.umap.merge$Species %in% c("Mouse.Bornstein"), ]
rownames(bornstein.cosine.meta) <- bornstein.cosine.meta$Sample

bornstein.cosine.pca <- prcomp_irlba(t(bornstein.cosine), n=50)

bornstein.batch.mnn <- reducedMNN(bornstein.cosine.pca$x,
                              batch=as.factor(bornstein.cosine.meta[colnames(bornstein.cosine), ]$CellType),
                              k=20)
```

Check the Bornstein batch correction worked OK - this corrects between experiments.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
bornstein.mnn.umap <- as.data.frame(umap(as.matrix(bornstein.batch.mnn$corrected[, c(1:50)]),
                               n_neighbors=7,
                               init='spectral',
                               metric='euclidean',
                               n_components=2,
                               min_dist=0.5,
                               method='naive')$layout)

colnames(bornstein.mnn.umap) <- paste0("MNN.Mouse.Teich.UMAP", c(1:2))
bornstein.mnn.umap$Sample <- colnames(bornstein.cosine)
bornstein.mnn.merge <- merge(bornstein.cosine.meta, bornstein.mnn.umap, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(bornstein.mnn.merge, aes(x=MNN.Mouse.Teich.UMAP1, y=MNN.Mouse.Teich.UMAP2, colour=CellType)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# MNN batch correction
cosine.bornstein.mouse <- cosineNorm(logcounts(combine.homolog.sce), subset.row=mouse.hvg.ensembl)
rownames(combine.umap.merge) <- combine.umap.merge$Sample

bornstein.mouse.mnn <- reducedMNN(cosine.pca$x,
                                   batch=as.factor(combine.umap.merge[colnames(cosine.bornstein.mouse), ]$Species),
                                   k=20)
bornstein.mouse.correct <- bornstein.mouse.mnn$corrected
rownames(bornstein.mouse.correct) <- combine.umap.merge[colnames(cosine.bornstein.mouse), ]$Sample
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
mnn.umap <- as.data.frame(umap(as.matrix(bornstein.mouse.correct[, c(1:20)]),
                               n_neighbors=15,
                               metric='euclidean',
                               n_components=2,
                               min_dist=0.5,
                               method='naive')$layout)

colnames(mnn.umap) <- paste0("MNN.UMAP", c(1:2))
mnn.umap$Sample <- rownames(bornstein.mouse.correct)
combine.mnn.merge <- merge(combine.umap.merge, mnn.umap, by='Sample')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(combine.mnn.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2, colour=Species)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_colorblind() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))

ggsave("~/Dropbox/AgeingExperiment/Human_Thymus/Bornstein_MNN_UMAP-Batch.png",
       height=5.15, width=9.95, dpi=300)
```

That looks pretty good - I hope the Bornstein-specific cells are their stromal (fibroblast) cells.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=9.95}
ggplot(combine.mnn.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2, colour=CellType)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=2, title.position="top", override.aes=list(size=3)))

ggsave("~/Dropbox/AgeingExperiment/Human_Thymus/bornstein_MNN_UMAP-TECtype.png",
       height=5.15, width=9.95, dpi=300)
```

Solved!!! They need to be in a joint-PC space.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
goi <- c("Ccl21a", "Krt5", "Aire", "Cd52", "Dclk1", "Avil", "Syngr1", "Chga", "Cd177", "Psmb11", "Prss16", "Krt80", "Spink5", "Lifr",
         "Ly6d", "Ly6e", "Ly6a", "Fgfr2", "Fgfr3", "Cdk1", "Pdpn", "Plekhb1", "Itgb2", "Dll4", "Cdh2", "Gas1", "Igfbp4", "Mgp", "Apoe", "Gsn", "Anxa1")
goi.ensembl <- gene.df$ensembl_gene_id[gene.df$external_gene_name %in% intersect(rownames(wildtype.sce), goi)]
mouse.goi.exprs <- as.data.frame(t(logcounts(mouse.sce[rownames(mouse.sce) %in% goi.ensembl, ])))
rownames(gene.df) <- gene.df$ensembl_gene_id
colnames(mouse.goi.exprs) <- gene.df[intersect(rownames(mouse.sce) , goi.ensembl), ]$external_gene_name 
mouse.goi.exprs$CellID <- colnames(mouse.sce)

combined.goi.exprs <- do.call(rbind.data.frame, list("teich"=goi.exprs, "mouse"=mouse.goi.exprs))
combine.goi.merge <- merge(combine.mnn.merge, combined.goi.exprs, by.x='Sample', by.y='CellID')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=11.95}
ggplot(combine.goi.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2)) +
  geom_point(data=combine.goi.merge[, c("MNN.UMAP1", "MNN.UMAP2")], 
             colour='grey80', alpha=0.5, size=0.25) +
  geom_point(size=0.5, aes(colour=Ccl21a)) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  facet_wrap(~Species) +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=11.95}
ggplot(combine.goi.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2)) +
  geom_point(data=combine.goi.merge[, c("MNN.UMAP1", "MNN.UMAP2")], 
             colour='grey80', alpha=0.5, size=0.25) +
  geom_point(size=0.5, aes(colour=Aire)) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  facet_wrap(~Species) +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=11.95}
ggplot(combine.goi.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2)) +
  geom_point(data=combine.goi.merge[, c("MNN.UMAP1", "MNN.UMAP2")], 
             colour='grey80', alpha=0.5, size=0.25) +
  geom_point(size=0.5, aes(colour=Spink5)) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  facet_wrap(~Species) +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=11.95}
ggplot(combine.goi.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2)) +
  geom_point(data=combine.goi.merge[, c("MNN.UMAP1", "MNN.UMAP2")], 
             colour='grey80', alpha=0.5, size=0.25) +
  geom_point(size=0.5, aes(colour=Dclk1)) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  facet_wrap(~Species) +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=11.95}
ggplot(combine.goi.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2)) +
  geom_point(data=combine.goi.merge[, c("MNN.UMAP1", "MNN.UMAP2")], 
             colour='grey80', alpha=0.5, size=0.25) +
  geom_point(size=0.5, aes(colour=Ly6a)) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  facet_wrap(~Species) +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```

I don't have any classes for the Bornstein - so I can only classify them into out TEC types, or the Teichman TEC types.

### kNN classification: Mouse subsets

I'll train on the mouse data and classify the bornstein cells first (k=5).

```{r, echo=FALSE, warning=FALSE, message=FALSE}
rownames(combine.goi.merge) <- combine.goi.merge$Sample
mouse.to.bornstein.knn <- knn(train=bornstein.mouse.correct[colnames(mouse.cosine), c(1:25)],
                               test=bornstein.mouse.correct[colnames(bornstein.cosine), c(1:25)],
                               cl=combine.goi.merge[colnames(mouse.cosine), ]$CellType,
                               k=5)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.knn <- data.frame("Sample"=colnames(bornstein.cosine), "knnClass"=mouse.to.bornstein.knn)
bornstein.knn.merge <- merge(bornstein.cosine.meta, bornstein.knn, by='Sample')

bornstein.class.props <- table(bornstein.knn.merge$knnClass, bornstein.knn.merge$CellType)
bornstein.class.props <- apply(bornstein.class.props, 2, FUN=function(X) X/sum(X))

pdf("~/Dropbox/AgeingExperiment/Human_Thymus/Mouse_to_Bornstein_confusion.pdf", height=3.95, width=3.95)
pheatmap(bornstein.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
dev.off()

pheatmap(bornstein.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

This heatmap shows the proportion of bornstein cells from each type (columns) that are classified into the different mouse groups by the kNN classifier. From this 
it looks like the mTEC(II) and mTEC(III) are closest to the Intertypical TEC.

I think I've got more confidence in the mouse cell type mapping than the bornstein. I'll try mapping the mouse into the bornstein classes anyway.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.to.mouse.knn <- knn(test=bornstein.mouse.correct[colnames(mouse.cosine), c(1:25)],
                          train=bornstein.mouse.correct[colnames(bornstein.cosine), c(1:25)],
                          cl=bornstein.cosine.meta[colnames(bornstein.cosine), ]$CellType,
                          k=5)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.knn <- data.frame("Sample"=colnames(mouse.cosine), "knnClass"=bornstein.to.mouse.knn)
mouse.knn.merge <- merge(combine.goi.merge, mouse.knn, by='Sample')

mouse.class.props <- table(mouse.knn.merge$knnClass, mouse.knn.merge$CellType)
mouse.class.props <- apply(mouse.class.props, 2, FUN=function(X) X/sum(X))

pdf("~/Dropbox/AgeingExperiment/Human_Thymus/bornstein_to_AgeMouse_confusion.pdf", height=3.95, width=3.95)
pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
dev.off()

pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

Now I've classified the mouse cells into the bornstein types - there is definitely less definition in the bornstein data set. The Intertypical TEC are smeared across the 
mTEC(I), mTEC(II) and mTEC(III) classes, the mature cTEC are also spread over the cTE, mcTEC and mTEC(I), the mature mTEC are split into mcTEC and mTEC(I). In 
fact, most of the cells are classified into mTEC(I) with the exception of the Tuft-like mTEC (which go into TEC(myo)?!), the perinatal cTEC which get split 
across the cTEC and mcTEC (at least this makes some sense) and the Post-Aire mTEC which are split between mTEC(I) and TEC(neuro).

Can I get the correct classification using the mouse data from this paper?

What if I batch correct everthing into the same space?

## Teichmann mouse data

```{r, echo=FALSE, warning=FALSE, message=FALSE}
postnatal.sce <- readRDS("~/Dropbox/AgeingExperiment/Human_Thymus/Postnatal_Teichmann-mouse_SCE.RDS")

teichmouse.meta <- read.table("~/Dropbox/AgeingExperiment/Human_Thymus/mouse_obs.csv",
                         sep="\t", header=TRUE, stringsAsFactors=FALSE)

teichmouse.genes <- rownames(postnatal.sce)

teichmouse.combine.meta <- teichmouse.meta[, c("index", "cell.types")]
colnames(teichmouse.combine.meta) <- c("Sample", "CellType")
teichmouse.combine.meta$Species <- "Mouse.Park"
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# map the teichmouse genes to ensembl IDs
rownames(match.gene.df) <- match.gene.df$external_gene_name
teichmouse.homolog.sce <- postnatal.sce[unique(match.gene.df$external_gene_name), ]
rownames(teichmouse.homolog.sce) <- match.gene.df[rownames(teichmouse.homolog.sce), ]$ensembl_gene_id
```

Combine all 3 data sets together.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
combine.homolog.sce <- do.call(cbind, list("mouse"=mouse.homolog.sce,
                                           "bornstein"=bornstein.homolog.sce,
                                           "teichmouse"=teichmouse.homolog.sce))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.all.meta <- do.call(rbind.data.frame, list("mouse"=mouse.combine.meta[, c("Sample", "CellType", "Species")],
                                                       "bornstein"=bornstein.combine.meta,
                                                       "park"=teichmouse.combine.meta))
rownames(mouse.all.meta) <- mouse.all.meta$Sample
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
cosine.all.mouse <- cosineNorm(logcounts(combine.homolog.sce), subset.row=mouse.hvg.ensembl)
cosine.pca <- prcomp_irlba(t(cosine.all.mouse), n=50)

all.mouse.mnn <- reducedMNN(cosine.pca$x,
                                   batch=as.factor(mouse.all.meta[colnames(cosine.all.mouse), ]$Species),
                                   k=20)
all.mouse.correct <- all.mouse.mnn$corrected
rownames(all.mouse.correct) <- mouse.all.meta[colnames(cosine.all.mouse), ]$Sample
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
mnn.umap <- as.data.frame(umap(as.matrix(all.mouse.correct[, c(1:20)]),
                               n_neighbors=15,
                               metric='euclidean',
                               n_components=2,
                               min_dist=0.5,
                               method='naive')$layout)

colnames(mnn.umap) <- paste0("MNN.UMAP", c(1:2))
mnn.umap$Sample <- rownames(all.mouse.correct)
combine.mnn.merge <- merge(mouse.all.meta, mnn.umap, by='Sample')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(combine.mnn.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2, colour=Species)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_colorblind() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))

ggsave("~/Dropbox/AgeingExperiment/Human_Thymus/All_MNN_UMAP-Batch.png",
       height=5.15, width=9.95, dpi=300)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
ggplot(combine.mnn.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2, colour=CellType)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  guides(colour=guide_legend(ncol=2, title.position="top", override.aes=list(size=3)))

ggsave("~/Dropbox/AgeingExperiment/Human_Thymus/All_MNN_UMAP-TECtype.png",
       height=5.15, width=9.95, dpi=300)
```

All 3 data sets map onto each other rather nicely.

### kNN classification: Mouse subsets

```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.cells <- combine.mnn.merge$Sample[combine.mnn.merge$Species %in% c("Mouse.Park")]
teichmouse.cosine <- cosine.all.mouse[, teichmouse.cells]
```

I'll train on the mouse data and classify the Teichmouse cells first (k=5).

```{r, echo=FALSE, warning=FALSE, message=FALSE}
rownames(combine.mnn.merge) <- combine.mnn.merge$Sample
mouse.to.teichmouse.knn <- knn(train=all.mouse.correct[colnames(mouse.cosine), c(1:25)],
                               test=all.mouse.correct[colnames(teichmouse.cosine), c(1:25)],
                               cl=combine.mnn.merge[colnames(mouse.cosine), ]$CellType,
                               k=5)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.knn <- data.frame("Sample"=colnames(teichmouse.cosine), "knnClass"=mouse.to.teichmouse.knn)
teichmouse.knn.merge <- merge(combine.mnn.merge, teichmouse.knn, by='Sample')

teichmouse.class.props <- table(teichmouse.knn.merge$knnClass, teichmouse.knn.merge$CellType)
teichmouse.class.props <- apply(teichmouse.class.props, 2, FUN=function(X) X/sum(X))

pdf("~/Dropbox/AgeingExperiment/Human_Thymus/Mouse_to_Teichmouse_confusion.pdf", height=3.95, width=3.95)
pheatmap(teichmouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
dev.off()

pheatmap(teichmouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

This heatmap shows the proportion of Park mouse cells from each type (columns) that are classified into the different mouse groups by the kNN 
classifier. From this it looks like the mTEC(II) and mTEC(III) are closest to the Intertypical TEC.

Now I'll classify the Bornstein cell using the Teichmouse.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.to.bornstein.knn <- knn(test=all.mouse.correct[colnames(bornstein.cosine), c(1:25)],
                                   train=all.mouse.correct[colnames(teichmouse.cosine), c(1:25)],
                                   cl=combine.mnn.merge[colnames(teichmouse.cosine), ]$CellType,
                                   k=5)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.knn <- data.frame("Sample"=colnames(bornstein.cosine), "knnClass"=teichmouse.to.bornstein.knn)
mouse.knn.merge <- merge(combine.mnn.merge, mouse.knn, by='Sample')

mouse.class.props <- table(mouse.knn.merge$knnClass, mouse.knn.merge$CellType)
mouse.class.props <- apply(mouse.class.props, 2, FUN=function(X) X/sum(X))

pdf("~/Dropbox/AgeingExperiment/Human_Thymus/Bornstein_to_TeichMouse_confusion.pdf", height=3.95, width=3.95)
pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
dev.off()

pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

Classify the Bornstein into Ageing TEC types.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.to.bornstein.knn <- knn(test=all.mouse.correct[colnames(bornstein.cosine), c(1:25)],
                                   train=all.mouse.correct[colnames(mouse.cosine), c(1:25)],
                                   cl=combine.mnn.merge[colnames(mouse.cosine), ]$CellType,
                                   k=5)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.knn <- data.frame("Sample"=colnames(bornstein.cosine), "knnClass"=mouse.to.bornstein.knn)
mouse.knn.merge <- merge(combine.mnn.merge, mouse.knn, by='Sample')

mouse.class.props <- table(mouse.knn.merge$knnClass, mouse.knn.merge$CellType)
mouse.class.props <- apply(mouse.class.props, 2, FUN=function(X) X/sum(X))

pdf("~/Dropbox/AgeingExperiment/Human_Thymus/Bornstein_to_AgeMouse_confusion.pdf", height=3.95, width=3.95)
pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
dev.off()

pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

Finally, classify ageing TEC into Teichman TEC types.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.to.mouse.knn <- knn(test=all.mouse.correct[colnames(mouse.cosine), c(1:25)],
                                   train=all.mouse.correct[colnames(teichmouse.cosine), c(1:25)],
                                   cl=combine.mnn.merge[colnames(teichmouse.cosine), ]$CellType,
                                   k=5)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.knn <- data.frame("Sample"=colnames(mouse.cosine), "knnClass"=teichmouse.to.mouse.knn)
mouse.knn.merge <- merge(combine.mnn.merge, mouse.knn, by='Sample')

mouse.class.props <- table(mouse.knn.merge$knnClass, mouse.knn.merge$CellType)
mouse.class.props <- apply(mouse.class.props, 2, FUN=function(X) X/sum(X))

pdf("~/Dropbox/AgeingExperiment/Human_Thymus/AgeMouse_to_TeichMouse_confusion.pdf", height=3.95, width=3.95)
pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
dev.off()

pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

Now I can compare the 3 data set classifications, using the Bornstein as the central one.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.class.df <- data.frame("Sample"=bornstein.cells, "Park.Class"=teichmouse.to.bornstein.knn,
                                 "Age.Class"=mouse.to.bornstein.knn, "Dataset"="Bornstein")

teichmouse.class.df <- data.frame("Sample"=teichmouse.knn.merge$Sample, "Park.Class"=teichmouse.knn.merge$CellType,
                                  "Age.Class"=teichmouse.knn.merge$knnClass, "Dataset"="Park")

agemouse.class.df <- data.frame("Sample"=mouse.knn.merge$Sample, "Park.Class"=mouse.knn.merge$knnClass,
                                  "Age.Class"=mouse.knn.merge$CellType, "Dataset"="Ageing")
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.class.props <- table(bornstein.class.df$Age.Class, bornstein.class.df$Park.Class)
bornstein.class.props <- apply(bornstein.class.props, 2, FUN=function(X) X/sum(X))

pdf("~/Dropbox/AgeingExperiment/Human_Thymus/Bornstein_ClassCompare_confusion.pdf", height=3.95, width=3.95)
pheatmap(bornstein.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
dev.off()

pheatmap(bornstein.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

Chris wants a riverplot for this. `ggalluvial` comes in handy here.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=9.95, fig.height=5.95}
# need to transform with xtabs first
bornstein.long <- as.data.frame(xtabs(data=bornstein.class.df, formula = ~ Park.Class + Age.Class))
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=14.95, fig.height=6.95}
# Stitch all datasets together
all.class.df <- do.call(rbind.data.frame, list("born"=bornstein.class.df, "park"=teichmouse.class.df, "age"=agemouse.class.df))
# count classes in each data set
all.class.long <- as.data.frame(xtabs(~ Age.Class + Park.Class + Dataset, data=all.class.df))

# set ordering of Age.Class
all.class.long$Age.Class <- factor(all.class.long$Age.Class,
                                   levels=c("sTEC", "Intertypical TEC", "nTEC", "Mature cTEC", "Perinatal cTEC",
                                            "Proliferating TEC", "Mature mTEC", "Post-Aire mTEC", "Tuft-like mTEC"),
                                   labels=c("sTEC", "Intertypical\nTEC", "nTEC", "Mature\ncTEC", "Perinatal\ncTEC",
                                             "Prolif.TEC", "Mature\nmTEC", "Post-Aire\nmTEC", "Tuft-like\nmTEC"))

all.class.long$Park.Class <- factor(all.class.long$Park.Class, 
                                    levels=c("TEC(neuro)_like_1", "cTEC", "cTEC(cycling)", "mTEC_I", "mTEC_II",
                                             "mTEC_III", "mTEC_IV(tuft)", "TEC(neuro)_like_2"),
                                    labels=c("TEC(neuro)1", "cTEC", "cTEC(cycling)", "mTEC(I)", "mTEC(II)",
                                             "mTEC(III)", "mTEC(IV)", "TEC(neuro)2"))

inter.cols <- c("#9970ab", "#35978f", "#B0cdc1", "#762a83", "#01665e", "#e7d4e8", "#dfc27d", "#8c510a" ,"#bf812d")
names(inter.cols) <- c("Post-Aire\nmTEC", 'Intertypical\nTEC', 'Mature\ncTEC', 'Tuft-like\nmTEC', 
                       'Prolif.TEC', 'Mature\nmTEC', 'nTEC', 'Perinatal\ncTEC', 'sTEC')


ggplot(all.class.long,
       aes(y=Freq, axis1=Age.Class, axis2=Park.Class)) +
  geom_alluvium(width=1/12, aes(fill=Age.Class)) +
  geom_stratum(width=1/12, colour='black', fill='grey80') +
  scale_fill_manual(values=inter.cols) +
  theme_mike() +
  theme(legend.position="right", legend.direction="vertical",
        axis.title=element_blank(), 
        panel.grid=element_blank(), axis.line=element_blank(),
        axis.text=element_blank(), axis.ticks=element_blank()) +
  guides(fill=guide_legend(title.position="top", ncol=1)) +
  geom_label(stat="stratum", infer.label=TRUE, size=2.9, angle=90) +
  facet_wrap(~Dataset, scales="free_y", nrow=1) +
  NULL

ggsave(filename="~/Dropbox/AgeingExperiment/Paper/eLife_submission/Classification_Riverplot.pdf",
      height=7.95, width=14.95, useDingbats=FALSE)


ggsave(filename="~/Dropbox/AgeingExperiment/Paper/eLife_submission/Classification_Riverplot.png",
      height=7.95, width=14.95, dpi=300)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=10.95, fig.height=5.95}
born.river <- ggplot(all.class.long[!all.class.long$Dataset %in% c("Bornstein"), ],
       aes(y=Freq, axis1=Age.Class, axis2=Park.Class)) +
  geom_alluvium(width=1/6, aes(fill=Age.Class)) +
  geom_stratum(width=1/6, colour='black', fill='grey80') +
  scale_fill_manual(values=inter.cols) +
  theme_mike() +
  theme(legend.position="right", legend.direction="vertical",
        strip.background=element_rect(fill='white', colour='white'),
        axis.title=element_blank(), 
        panel.grid=element_blank(), axis.line=element_blank(),
        axis.text=element_blank(), axis.ticks=element_blank()) +
  geom_label(stat="stratum", infer.label=TRUE, size=2.9, angle=90, fill='white') +
  guides(fill=guide_legend(title.position="top", ncol=1),
         text=FALSE) +
  facet_wrap(~Dataset, scales="free_y", nrow=1) +
  NULL

ggsave(filename="~/Dropbox/AgeingExperiment/Paper/eLife_submission/Classification_Riverplot_sansBornstein.png",
      height=5.95, width=10.95, dpi=300)

born.river
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
reiver.big <- ggplot(all.class.long[!all.class.long$Dataset %in% c("Bornstein"), ],
       aes(y=Freq, axis1=Age.Class, axis2=Park.Class)) +
  geom_alluvium(width=1/6, aes(fill=Age.Class)) +
  geom_stratum(width=1/6, colour='black', fill='grey80') +
  scale_fill_manual(values=inter.cols) +
  theme_mike() +
  theme(legend.position="right", legend.direction="vertical",
        strip.background=element_rect(fill='white', colour='white'),
        axis.title=element_blank(), 
        legend.text=element_text(size=24), legend.title=element_text(size=32, face="bold"),
        strip.text=element_text(size=36),
        panel.grid=element_blank(), axis.line=element_blank(),
        axis.text=element_blank(), axis.ticks=element_blank()) +
  geom_label(stat="stratum", infer.label=TRUE, size=5.25, angle=90, fill='white', fontface="bold") +
  guides(fill=guide_legend(title.position="top", title="Ageing Label", 
                           ncol=1),
         text=FALSE) +
  facet_wrap(~Dataset, scales="free_y", nrow=1) +
  NULL

ggsave(reiver.big,
       filename="~/Dropbox/AgeingExperiment/Paper/eLife_submission/Classification_Riverplot_sansBornstein.pdf",
       height=11.95, width=21.95, useDingbats=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=10.95, fig.height=12.95}
ggplot(all.class.long[!all.class.long$Dataset %in% c("Bornstein"), ],
       aes(y=Freq, axis1=Age.Class, axis2=Dataset)) +
  geom_alluvium(width=1/6, aes(fill=Dataset)) +
  geom_stratum(width=1/6, colour='black', fill='grey80') +
  scale_fill_manual(values=c("orange", "blue")) +
  theme_mike() +
  theme(legend.position="right", legend.direction="vertical",
        strip.background=element_rect(fill='white', colour='white'),
        strip.text=element_text(size=10, face='bold'),
        axis.title=element_blank(), 
        panel.grid=element_blank(), axis.line=element_blank(),
        axis.text=element_blank(), axis.ticks=element_blank()) +
  guides(fill=guide_legend(title.position="top", ncol=1),
         text=FALSE) +
  facet_grid(Park.Class~Age.Class, scales="free") +
  NULL

ggsave(filename="~/Dropbox/AgeingExperiment/Paper/eLife_submission/Classification_Riverplot_sansBornstein-nolabel.pdf",
      height=12.95, width=10.95, useDingbats=FALSE)

ggsave(filename="~/Dropbox/AgeingExperiment/Paper/eLife_submission/Classification_Riverplot_sansBornstein-nolabel.png",
      height=12.95, width=10.95, dpi=300)
```

How are the classifications distributed over the original experimental samples, i.e. age, donors, etc?

```{r, echo=FALSE, warning=FALSE, message=FALSE}
bornstein.class.merge <- merge(bornstein.class.df, wildtype.homolog.umap, by.x='Sample', by.y='well')

ggplot(bornstein.class.merge, aes(x=Age.Class, fill=Experiment_ID)) +
  geom_bar(position='fill') +
  scale_fill_colorblind() +
  theme_mike() +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1),
        legend.position="bottom", legend.direction="horizontal",
        legend.text=element_text(size=10), legend.title=element_text(size=10)) +
  labs(x="Ageing Classification", y="Proportion") +
  guides(fill=guide_legend())

ggsave("~/Dropbox/AgeingExperiment/Paper/eLife_submission/Bornstein_AgeVsAgeClass-bar.pdf",
       height=4.25, width=5.95, useDingbats=FALSE)

ggsave("~/Dropbox/AgeingExperiment/Paper/eLife_submission/Bornstein_AgeVsAgeClass-bar.png",
       height=4.25, width=5.95, dpi=300)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
park.class.merge <- merge(teichmouse.class.df, teichmouse.meta, by.x='Sample', by.y='index')

ggplot(park.class.merge, aes(x=Age.Class, fill=age)) +
  geom_bar(position='fill') +
  scale_fill_colorblind() +
  theme_mike() +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1),
        legend.position="bottom", legend.direction="horizontal",
        legend.text=element_text(size=10), legend.title=element_text(size=10)) +
  labs(x="Ageing Classification", y="Proportion") +
  guides(fill=guide_legend(title="Age"))

ggsave("~/Dropbox/AgeingExperiment/Paper/eLife_submission/Park_AgeVsAgeClass-bar.pdf",
       height=4.25, width=5.95, useDingbats=FALSE)

ggsave("~/Dropbox/AgeingExperiment/Paper/eLife_submission/Park_AgeVsAgeClass-bar.png",
       height=4.25, width=5.95, dpi=300)
```




