---
title: "Comparing mouse and mouse thymus"
output: html_notebook
---

# Introduction

Comparing our ZsG 10X mouse thymus data with mouse TEC from Park _et al_ Science 2020. I need to do this on just the TEC cells.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(igraph)
library(SingleCellExperiment)
library(scran)
library(scater)
library(Matrix)
library(umap)
library(irlba)
library(scales)
library(viridis)
library(ggplot2)
library(biomaRt)
library(ggthemes)
library(batchelor)
library(FNN)
library(pheatmap)
source("~/Dropbox/R_sessions/SingleCellFunctions/single_cell_functions.R")
source("~/Dropbox/R_sessions/GGMike/theme_mike.R")
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
postnatal.sce <- readRDS(file="~/Dropbox/AgeingExperiment/Human_Thymus/Postnatal_Teichmann-mouse_SCE.RDS")
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.meta <- read.table("~/Dropbox/AgeingExperiment/Human_Thymus/mouse_obs.csv",
                         sep="\t", header=TRUE, stringsAsFactors=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
postnatal.cells <- teichmouse.meta$index[teichmouse.meta$age %in% c("6 days", "4-6 weeks")  & (grepl(teichmouse.meta$cell.types, pattern="TEC"))]

postnatal.hvg <- find_hvg(postnatal.sce, p.threshold=1e-3, return.ranks=TRUE)
table(postnatal.hvg$HVG)
```

The bulk of cells are from post-natal thymi.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
postnatal.pca <- prcomp_irlba(as.matrix(t(logcounts(postnatal.sce[postnatal.hvg$HVG, ]))), n=50)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
postnatal.umap <- as.data.frame(umap(as.matrix(postnatal.pca$x[, c(1:20)]),
                                     n_neighbors=11,
                                     init='spectral',
                                     metric='cosine',
                                     n_components=2,
                                     min_dist=0.1,
                                     method='naive')$layout)

colnames(postnatal.umap) <- paste0("UMAP", c(1:2))
postnatal.umap$CellID <- colnames(postnatal.sce)

postnatal.meta.umap <- merge(teichmouse.meta, postnatal.umap, by.x='index', by.y='CellID')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(postnatal.meta.umap, aes(x=UMAP1, y=UMAP2, colour=cell.types)) +
  geom_point(size=0.1) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(override.aes=list(size=3), ncol=1))
```

This includes fibroblasts and non-TEC cells. I'll leave these in as a control for the comparison with our cells.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(postnatal.meta.umap, aes(x=UMAP1, y=UMAP2, colour=age)) +
  geom_point(size=0.1) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(override.aes=list(size=3), ncol=1))
```

Hmm, are these batch effects or genuine age-related differences? There's a fair bit of mixing, and there aren't a huge number of 6 day cells.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
goi <- c("Ccl21a", "Krt5", "Aire", "Cd52", "Dclk1", "Avil", "Syngr1", "Chga", "Cd177", "Psmb11", "Prss16", "Krt80", "Spink5", "Lifr",
         "Ly6d", "Ly6e", "Ly6a", "Fgfr2", "Fgfr3", "Cdk1", "Pdpn", "Plekhb1", "Itgb2", "Dll4", "Cdh2", "Gas1", "Igfbp4", "Mgp", "Apoe", "Gsn", "Anxa1")
goi.exprs <- as.data.frame(t(logcounts(postnatal.sce[goi, ])))
goi.exprs$CellID <- colnames(postnatal.sce)

postnatal.goi.merge <- merge(postnatal.meta.umap, goi.exprs, by.x='index', by.y='CellID')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(postnatal.goi.merge, aes(x=UMAP1, y=UMAP2, colour=Ccl21a)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_colourbar(ncol=1, titke.position="top"))
```

Based on _Ccl21a_ it looks like the intertypical TEC are in amongst the mTEC(I) population again.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(postnatal.goi.merge, aes(x=UMAP1, y=UMAP2, colour=Lifr)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_colourbar(ncol=1, titke.position="top"))
```

_Lifr_ is also a bit sporadic but overlaps mostly with the mTEC(I) that also express _Ccl21a_ - are the mTEC(I) a misclassification?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(postnatal.goi.merge, aes(x=UMAP1, y=UMAP2, colour=Ly6a)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_colourbar(ncol=1, titke.position="top"))
```

This is _Ly6a_, again note the overlap - I think the Intertypical TEC are a subset of the mTEC(I) population.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(postnatal.goi.merge, aes(x=UMAP1, y=UMAP2, colour=Pdpn)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_colourbar(ncol=1, titke.position="top"))
```

_Pdpn_ is most evident in the fibroblasts, but some sporadic expression through the mTEC(I) compartment as well.

Let's try something a bit different, namely use our mouse thymus HVGs to define the Teichmann mouse TEC manifold. 

```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.hvgs <- read.table("~/Dropbox/AgeingExperiment/Thymus_HVG.tsv",
                         sep="\t", header=TRUE, stringsAsFactors=FALSE)
table(mouse.hvgs$HVG)
```

There are just shy of 5000 mouse HVGs. I'll use these to define the manifold in the Teichmann mouse cells.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
biomaRt.connection <- useMart("ensembl", "mmusculus_gene_ensembl")

gene.df <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "chromosome_name"),
                 mart=biomaRt.connection)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.hvg.teichmouse <- intersect(rownames(postnatal.sce), gene.df$external_gene_name[gene.df$ensembl_gene_id %in% mouse.hvgs$gene_id[mouse.hvgs$HVG]])

mouse.homolog.pca <- prcomp_irlba(as.matrix(t(logcounts(postnatal.sce[mouse.hvg.teichmouse, ]))), n=50)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
mouse.homolog.umap <- as.data.frame(umap(as.matrix(mouse.homolog.pca$x[, c(1:20)]),
                                     n_neighbors=7,
                                     init='spectral',
                                     metric='cosine',
                                     n_components=2,
                                     min_dist=0.75,
                                     method='naive')$layout)

colnames(mouse.homolog.umap) <- paste0("Homolog.UMAP", c(1:2))
mouse.homolog.umap$CellID <- colnames(postnatal.sce)

postnatal.homolog.umap <- merge(postnatal.goi.merge, mouse.homolog.umap, by.x='index', by.y='CellID')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(postnatal.homolog.umap, aes(x=Homolog.UMAP1, y=Homolog.UMAP2, colour=cell.types)) +
  geom_point(size=0.1) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(override.aes=list(size=3), ncol=1))
```

Our mouse HVGs still capture a lot of the variance in the Teichmann data as well. Interesting the mTEC(I) population starts to splinter.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(postnatal.homolog.umap, aes(x=Homolog.UMAP1, y=Homolog.UMAP2, colour=age)) +
  geom_point(size=0.1) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(override.aes=list(size=3), ncol=1))
```


OK, I need to do this concretely and map the teichmouse and our mouse data together.

## Combining teichmouse and our mouse data

```{r, echo=FALSE, warning=FALSE, message=FALSE}
thymus.meta <- read.table("~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ZsG_GOI_metadata.tsv",
                       sep="\t", header=TRUE, stringsAsFactors=FALSE)

n.clusters <- length(unique(thymus.meta$ZsG.Cluster))

thymus.meta$TECtype <- "Unknown"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(1)] <- "mTEC.1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(3)] <- "mTEC.2"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(14)] <- "mTEC.3"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(6)] <- "mTEC.4"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(20)] <- "mTEC.5"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(22)] <- "mTEC.6"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(16)] <- "mTEC.7"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(4)] <- "Intertypical.TEC.1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(7)] <- "Intertypical.TEC.2"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(10)] <- "Intertypical.TEC.3"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(11)] <- "Intertypical.TEC.4"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(9)] <- "cTEC.1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(13)] <- "cTEC.2"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(17)] <- "PostAire.mTEC.1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(19)] <- "PostAire.mTEC.2"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(23)] <- "eTEC1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(24)] <- "Prolif.TEC.1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(12)] <- "Prolif.TEC.2"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(2)] <- "Prolif.TEC.3"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(15)] <- "Tuft.mTEC.1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(18)] <- "Tuft.mTEC.2"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(5)] <- "Sca1.TEC.1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(8)] <- "New.TEC.1"
thymus.meta$TECtype[thymus.meta$ZsG.Cluster %in% c(21)] <- "New.TEC.2"
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.sce <- readRDS("~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ALL_SCE.RDS")
```

First I'll combine the teichmouse and mouse SCE objects using a set of commonly expressed 1:1 homologous genes.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.genes <- rownames(mouse.sce)
teichmouse.genes <- rownames(postnatal.sce)

match.gene.df <- gene.df[gene.df$external_gene_name %in% teichmouse.genes & gene.df$ensembl_gene_id %in% mouse.genes, ]
match.gene.df <- match.gene.df[!duplicated(match.gene.df$external_gene_name), ]
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# map the teichmouse genes to ensembl IDs
rownames(match.gene.df) <- match.gene.df$external_gene_name
teichmouse.homolog.sce <- postnatal.sce[unique(match.gene.df$external_gene_name), ]
rownames(teichmouse.homolog.sce) <- match.gene.df[rownames(teichmouse.homolog.sce), ]$ensembl_gene_id

mouse.homolog.sce <- mouse.sce[unique(match.gene.df$ensembl_gene_id), ]
rownames(match.gene.df) <- match.gene.df$ensembl_gene_id
rownames(mouse.homolog.sce) <- match.gene.df[rownames(mouse.homolog.sce), ]$ensembl_gene_id

combine.homolog.sce <- do.call(cbind, list("mouse"=mouse.homolog.sce, "teichmouse"=teichmouse.homolog.sce))
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# make a combined meta-data
teichmouse.combine.meta <- postnatal.goi.merge[, c("index", "cell.types", "age", "sample_ID")]

rownames(thymus.meta) <- thymus.meta$Sample
mouse.combine.meta <- thymus.meta[colnames(mouse.homolog.sce), c("Sample", "TECtype", "Age", "SampID")]

colnames(teichmouse.combine.meta) <- c("Sample", "CellType", "Age", "Donor")
colnames(mouse.combine.meta) <- c("Sample", "CellType", "Age", "Donor")
teichmouse.combine.meta$Species <- "Mouse.Teich"
mouse.combine.meta$Species <- "Mouse.ZsG"

mouse.teichmouse.meta <- do.call(rbind.data.frame, list("teichmouse"=teichmouse.combine.meta,
                                                   "mouse"=mouse.combine.meta[, c("Sample", "CellType", 
                                                                                  "Donor", "Age", "Species")]))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
rownames(match.gene.df) <- match.gene.df$external_gene_name
mouse.hvg.ensembl <- match.gene.df[mouse.hvg.teichmouse, ]$ensembl_gene_id
mouse.hvg.ensembl <- mouse.hvg.ensembl[!is.na(mouse.hvg.ensembl)]
combine.homolog.pca <- prcomp_irlba(as.matrix(t(logcounts(combine.homolog.sce[mouse.hvg.ensembl, ]))), n=50)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
combine.homolog.umap <- as.data.frame(umap(as.matrix(combine.homolog.pca$x[, c(1:20)]),
                                           n_neighbors=7,
                                           init='spectral',
                                           metric='cosine',
                                           n_components=2,
                                           min_dist=0.75,
                                           method='naive')$layout)

colnames(combine.homolog.umap) <- paste0("Combine.UMAP", c(1:2))
combine.homolog.umap$Sample <- colnames(combine.homolog.sce)
combine.umap.merge <- merge(mouse.teichmouse.meta, combine.homolog.umap, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(combine.umap.merge, aes(x=Combine.UMAP1, y=Combine.UMAP2, colour=Species)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_colorblind() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, titke.position="top", override.aes=list(size=3)))
```

OK - so firstly there is a clear batch effect that needs to be dealt with.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=8.95}
ggplot(combine.umap.merge, aes(x=Combine.UMAP1, y=Combine.UMAP2, colour=CellType)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=2, titke.position="top", override.aes=list(size=3)))
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(combine.umap.merge, aes(x=Combine.UMAP1, y=Combine.UMAP2, colour=Age)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, titke.position="top", override.aes=list(size=3)))
```

The ages are much more mixed in the mouse than the teichmouse - is this biological or technical though? I worry the latter. I'll batch correct 
and re-run the UMAP.

### Teichmann - mouse batch correction

The first step is a cosine normalisation of the gene expression before PCA.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
cosine.teichmouse.mouse <- cosineNorm(logcounts(combine.homolog.sce), subset.row=mouse.hvg.ensembl)
cosine.pca <- prcomp_irlba(t(cosine.teichmouse.mouse), n=50)
```

I'll then batch correct the teichmouse data together.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.cells <- combine.umap.merge$Sample[combine.umap.merge$Species %in% c("Mouse.Teich")]
mouse.cells <- combine.umap.merge$Sample[combine.umap.merge$Species %in% c("Mouse.ZsG")]

teichmouse.cosine <- cosine.teichmouse.mouse[, teichmouse.cells]
mouse.cosine <- cosine.teichmouse.mouse[, mouse.cells]
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.cosine.meta <- combine.umap.merge[combine.umap.merge$Species %in% c("Mouse.Teich"), ]
rownames(teichmouse.cosine.meta) <- teichmouse.cosine.meta$Sample

teichmouse.cosine.pca <- prcomp_irlba(t(teichmouse.cosine), n=50)

teichmouse.batch.mnn <- reducedMNN(teichmouse.cosine.pca$x,
                              batch=as.factor(teichmouse.cosine.meta[colnames(teichmouse.cosine), ]$Donor),
                              k=20)
```

Check the teichmouse batch correction worked OK

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
teichmouse.mnn.umap <- as.data.frame(umap(as.matrix(teichmouse.batch.mnn$corrected[, c(1:50)]),
                               n_neighbors=7,
                               init='spectral',
                               metric='euclidean',
                               n_components=2,
                               min_dist=0.5,
                               method='naive')$layout)

colnames(teichmouse.mnn.umap) <- paste0("MNN.Mouse.Teich.UMAP", c(1:2))
teichmouse.mnn.umap$Sample <- colnames(teichmouse.cosine)
teichmouse.mnn.merge <- merge(teichmouse.cosine.meta, teichmouse.mnn.umap, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(teichmouse.mnn.merge, aes(x=MNN.Mouse.Teich.UMAP1, y=MNN.Mouse.Teich.UMAP2, colour=CellType)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(teichmouse.mnn.merge, aes(x=MNN.Mouse.Teich.UMAP1, y=MNN.Mouse.Teich.UMAP2, colour=Age)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_colorblind() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))
```

That's better. There is still some age-specific effects retained after batch correction. I'll do the same for our mouse then combine these PCs together.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.cosine.meta <- combine.umap.merge[combine.umap.merge$Species %in% c("Mouse.ZsG"), ]
rownames(mouse.cosine.meta) <- mouse.cosine.meta$Sample

mouse.cosine.pca <- prcomp_irlba(t(mouse.cosine), n=50)

mouse.batch.mnn <- reducedMNN(mouse.cosine.pca$x,
                              batch=as.factor(mouse.cosine.meta[colnames(mouse.cosine), ]$Donor),
                              k=20)
```

Check the teichmouse batch correction worked OK

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
mouse.mnn.umap <- as.data.frame(umap(as.matrix(mouse.batch.mnn$corrected[, c(1:50)]),
                               n_neighbors=11,
                               init='spectral',
                               metric='euclidean',
                               n_components=2,
                               min_dist=0.5,
                               method='naive')$layout)

colnames(mouse.mnn.umap) <- paste0("MNN.Mouse.UMAP", c(1:2))
mouse.mnn.umap$Sample <- colnames(mouse.cosine)
mouse.mnn.merge <- merge(mouse.cosine.meta, mouse.mnn.umap, by='Sample')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(mouse.mnn.merge, aes(x=MNN.Mouse.UMAP1, y=MNN.Mouse.UMAP2, colour=CellType)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(mouse.mnn.merge, aes(x=MNN.Mouse.UMAP1, y=MNN.Mouse.UMAP2, colour=Age)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_colorblind() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))
```



```{r, echo=FALSE, warning=FALSE, message=FALSE}
# MNN batch correction
teichmouse.mouse.mnn <- reducedMNN(teichmouse.batch.mnn$corrected, mouse.batch.mnn$corrected,
                              k=31)
teichmouse.mouse.correct <- teichmouse.mouse.mnn$corrected
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
mnn.umap <- as.data.frame(umap(as.matrix(teichmouse.mouse.correct[, c(1:10)]),
                               n_neighbors=15,
                               metric='euclidean',
                               n_components=2,
                               min_dist=0.1,
                               method='naive')$layout)

colnames(mnn.umap) <- paste0("MNN.UMAP", c(1:2))
mnn.umap$Sample <- c(colnames(teichmouse.cosine), colnames(mouse.cosine))
combine.mnn.merge <- merge(combine.umap.merge, mnn.umap, by='Sample')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=7.25}
ggplot(combine.mnn.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2, colour=Species)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_colorblind() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=1, title.position="top", override.aes=list(size=3)))
```

Hmm, still not as well mixed as I would have expected.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=9.95}
ggplot(combine.mnn.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2, colour=CellType)) +
  geom_point(size=0.25) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical") +
  guides(colour=guide_legend(ncol=2, titke.position="top", override.aes=list(size=3)))
```

Maybe I'll try this in gene expression space, then do the PCA?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=11.95}
ggplot(combine.mnn.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2)) +
  geom_point(data=combine.mnn.merge[, c("MNN.UMAP1", "MNN.UMAP2")], 
             colour='grey80', alpha=0.5, size=0.25) +
  geom_point(size=0.5, aes(colour=CellType)) +
  theme_mike() +
  scale_colour_Publication() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  facet_wrap(~Species) +
  guides(colour=guide_legend(ncol=2, titke.position="top", override.aes=list(size=3)))
```

Umm, I'm concerned that their mapping is terribly, terribly wrong. Which of these is the gold standard? I would say ours as we sorted the cTEC and mTEC, etc 
separately, but this is also a poor indicator of actual provenance/lineage. Generally though, the link from Intertypical TEC to mTEC isn't preserved well here. 
I think I should compare to our 10X data.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
goi <- c("Ccl21a", "Krt5", "Aire", "Cd52", "Dclk1", "Avil", "Syngr1", "Chga", "Cd177", "Psmb11", "Prss16", "Krt80", "Spink5", "Lifr",
         "Ly6d", "Ly6e", "Ly6a", "Fgfr2", "Fgfr3", "Cdk1", "Pdpn", "Plekhb1", "Itgb2", "Dll4", "Cdh2", "Gas1", "Igfbp4", "Mgp", "Apoe", "Gsn", "Anxa1")
goi.ensembl <- gene.df$ensembl_gene_id[gene.df$external_gene_name %in% goi]
mouse.goi.exprs <- as.data.frame(t(logcounts(mouse.sce[rownames(mouse.sce) %in% goi.ensembl, ])))
rownames(gene.df) <- gene.df$ensembl_gene_id
colnames(mouse.goi.exprs) <- gene.df[intersect(rownames(mouse.sce) , goi.ensembl), ]$external_gene_name 
mouse.goi.exprs$CellID <- colnames(mouse.sce)

combined.goi.exprs <- do.call(rbind.data.frame, list("teich"=goi.exprs, "mouse"=mouse.goi.exprs))
combine.goi.merge <- merge(combine.mnn.merge, combined.goi.exprs, by.x='Sample', by.y='CellID')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=11.95}
ggplot(combine.goi.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2)) +
  geom_point(data=combine.goi.merge[, c("MNN.UMAP1", "MNN.UMAP2")], 
             colour='grey80', alpha=0.5, size=0.25) +
  geom_point(size=0.5, aes(colour=Ccl21a)) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  facet_wrap(~Species) +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.15, fig.width=11.95}
ggplot(combine.goi.merge, aes(x=MNN.UMAP1, y=MNN.UMAP2)) +
  geom_point(data=combine.goi.merge[, c("MNN.UMAP1", "MNN.UMAP2")], 
             colour='grey80', alpha=0.5, size=0.25) +
  geom_point(size=0.5, aes(colour=Aire)) +
  theme_mike() +
  scale_colour_viridis() +
  theme(legend.position="right", legend.direction="vertical",
        legend.text=element_text(size=11)) +
  facet_wrap(~Species) +
  guides(colour=guide_colourbar(ncol=1, title.position="top"))
```




This shows the same plots by age - most of the overlap is with adult teichmouse TEC - and actually the 3month old cells are relaly different to the rest. Perhaps I 
can try this on just the adult teichmouse TEC? I'll try the classification approach using kNNs, a Random Forest and a multinomial model.

### kNN classification: Mouse subsets

I'll train on the mouse data and classify the teichmouse cells first (k=5).

```{r, echo=FALSE, warning=FALSE, message=FALSE}
rownames(teichmouse.mouse.correct) <- c(colnames(teichmouse.cosine), colnames(mouse.cosine))

mouse.to.teichmouse.knn <- knn(train=teichmouse.mouse.correct[colnames(mouse.cosine), c(1:25)],
                          test=teichmouse.mouse.correct[colnames(teichmouse.cosine), c(1:25)],
                          cl=mouse.cosine.meta[colnames(mouse.cosine), ]$CellType,
                          k=5)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.knn <- data.frame("Sample"=colnames(teichmouse.cosine), "knnClass"=mouse.to.teichmouse.knn)
teichmouse.knn.merge <- merge(teichmouse.cosine.meta, teichmouse.knn, by='Sample')

teichmouse.class.props <- table(teichmouse.knn.merge$knnClass, teichmouse.knn.merge$CellType)
teichmouse.class.props <- apply(teichmouse.class.props, 2, FUN=function(X) X/sum(X))
pheatmap(teichmouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

This heatmap shows the proportion of teichmouse cells from each type (columns) that are classified into the different mouse groups by the kNN classifier. From this 
it looks like the mTEC(II) and mTEC(III) are closest to the Intertypical TEC.

As a sanity check, how does this perform cross-mapping over ages in both teichmouse and mouse?

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# train on 19 and classify on 35yo
teichmouse.6dayso.cells <- teichmouse.cosine.meta$Sample[teichmouse.cosine.meta$Age %in% c("6 days")]
teichmouse.4.6wks.cells  <- teichmouse.cosine.meta$Sample[teichmouse.cosine.meta$Age %in% c("4-6 weeks")]

teichmouse.ages.knn <- knn(train=teichmouse.mouse.correct[teichmouse.6dayso.cells, c(1:25)],
                      test=teichmouse.mouse.correct[teichmouse.4.6wks.cells, c(1:25)],
                      cl=teichmouse.cosine.meta[teichmouse.6dayso.cells, ]$CellType,
                      k=21)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
teichmouse.age.knn <- data.frame("Sample"=teichmouse.4.6wks.cells, "knnClass"=teichmouse.ages.knn)
teichmouse.age.knn.merge <- merge(teichmouse.cosine.meta, teichmouse.age.knn, by='Sample')

teichmouse.age.class.props <- table(teichmouse.age.knn.merge$knnClass, teichmouse.age.knn.merge$CellType)
teichmouse.age.class.props <- apply(teichmouse.age.class.props, 2, FUN=function(X) X/sum(X))
pheatmap(teichmouse.age.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

Even within their own data this does poorly - are they definitely fibrolasts?!

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# train on 19 and classify on 4.6wks
mouse.4wk.cells <- mouse.cosine.meta$Sample[mouse.cosine.meta$Age %in% c("4wk")]
mouse.16wk.cells  <- mouse.cosine.meta$Sample[mouse.cosine.meta$Age %in% c("16wk")]

mouse.ages.knn <- knn(train=teichmouse.mouse.correct[mouse.4wk.cells, c(1:25)],
                      test=teichmouse.mouse.correct[mouse.16wk.cells, c(1:25)],
                      cl=mouse.cosine.meta[mouse.4wk.cells, ]$CellType,
                      k=21)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.age.knn <- data.frame("Sample"=mouse.16wk.cells, "knnClass"=mouse.ages.knn)
mouse.age.knn.merge <- merge(mouse.cosine.meta, mouse.age.knn, by='Sample')

mouse.age.class.props <- table(mouse.age.knn.merge$knnClass, mouse.age.knn.merge$CellType)
mouse.age.class.props <- apply(mouse.age.class.props, 2, FUN=function(X) X/sum(X))
pheatmap(mouse.age.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

There's some error in the sTEC which get classified into Intertpical and mature cTEC (there are only 3 sTEC though) - likeiwse for the 2 perinatal cTEC, which are
classified into mature cTEC, and the 8 nTEC which are classified into either nTEC or mature mTEC.

I think I've got more confidence in the mouse cell type mapping than the teichmouse. I'll try mapping the mouse into the teichmouse classes anyway.


```{r, echo=FALSE, warning=FALSE, message=FALSE}
rownames(teichmouse.mouse.correct) <- c(colnames(teichmouse.cosine), colnames(mouse.cosine))

teichmouse.to.mouse.knn <- knn(test=teichmouse.mouse.correct[colnames(mouse.cosine), c(1:25)],
                          train=teichmouse.mouse.correct[colnames(teichmouse.cosine), c(1:25)],
                          cl=teichmouse.cosine.meta[colnames(teichmouse.cosine), ]$CellType,
                          k=5)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
mouse.knn <- data.frame("Sample"=colnames(mouse.cosine), "knnClass"=teichmouse.to.mouse.knn)
mouse.knn.merge <- merge(mouse.cosine.meta, mouse.knn, by='Sample')

mouse.class.props <- table(mouse.knn.merge$knnClass, mouse.knn.merge$CellType)
mouse.class.props <- apply(mouse.class.props, 2, FUN=function(X) X/sum(X))
pheatmap(mouse.class.props, cluster_cols=FALSE, cluster_rows=FALSE)
```

Now I've classified the mouse cells into the teichmouse types - there is definitely less definition in the teichmouse data set. The Intertypical TEC are smeared across the 
mTEC(I), mTEC(II) and mTEC(III) classes, the mature cTEC are also spread over the cTE, mcTEC and mTEC(I), the mature mTEC are split into mcTEC and mTEC(I). In 
fact, most of the cells are classified into mTEC(I) with the exception of the Tuft-like mTEC (which go into TEC(myo)?!), the perinatal cTEC which get split 
across the cTEC and mcTEC (at least this makes some sense) and the Post-Aire mTEC which are split between mTEC(I) and TEC(neuro).

Can I get the correct classification using the mouse data from this paper?



