---
title: "Thymus ZsG 10X: Subsample analysis"
output: NA
---

# Introduction

I've now run the separate graph-construction, clustering, etc on the full 69,000 QC'd single cells. I'll do the visualisations here.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(ggplot2)
library(scales)
library(biomaRt)
library(ggrepel)
library(scran)
library(irlba)
library(stringr)
library(igraph)
library(ggrepel)
library(reshape2)
library(umap)
library(pheatmap)
library(ComplexHeatmap)
library(viridis)
library(FNN)
library(cowplot)
library(batchelor)
library(schex)
library(ggrastr)
library(cydar)
library(edgeR)
library(RColorBrewer)
library(destiny)
source("~/Dropbox/R_sessions/SingleCellFunctions/single_cell_functions.R")
source("~/Dropbox/R_sessions/GGMike/theme_mike.R")
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
gene.df <- read.table("~/CI_filesystem/mnt/scratchb/jmlab/morgan02/10X/Thymus_ZsG/ZsG_genes-biomart.tsv",
                      sep="\t", header=TRUE, stringsAsFactors=FALSE)
rownames(gene.df) <- gene.df$ensembl_gene_id
```

I tested for HVGs with FDR $\lt 10^{-7}$.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
zsg.hvg <- read.table("~/CI_filesystem/mnt/scratchb/jmlab/morgan02/10X/Thymus_ZsG/ZsG_HVGs.tsv",
                      sep="\t", header=TRUE, stringsAsFactors=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
zsg.meta <- read.table("~/CI_filesystem/mnt/scratchb/jmlab/morgan02/10X/Thymus_ZsG/ZsG_GOI_metadata.tsv",
                       sep="\t", header=TRUE, stringsAsFactors=FALSE)

n.clusters <- length(unique(zsg.meta$ZsG.Cluster))
zsg.meta$ZsG.Cluster <- factor(zsg.meta$ZsG.Cluster,
                               levels=c(1:n.clusters),
                               labels=c(1:n.clusters))

zsg.meta$Age <- factor(zsg.meta$Age,
                               levels=c("Wk1", "Wk4", "Wk16"),
                               labels=c("Wk1", "Wk4", "Wk16"))
```

Firstly, how many clusters do we find, and how are they distributed across the experimental samples and SNN-graph.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
table(zsg.meta$ZsG.Cluster)
```

There are 24 clusters - as with the sub-sample analysis I would not be surprised if we need to merge these to get biologically coherent TEC subtypes.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
zsg.cluster.cols <- c("#3E89BE", "#fdb462", "#7fc97f","#ef3b2c","#662506",
                     "#a6cee3","#fb9a99","#984ea3", "#ffff33", "#b2df8a",
                     "#006FFF", "#CA834E", "#518A87", "#5B113C", "#55813B",
                     "#E704C4", "#DDAEA2",  "#77837F", "#b15928",
                     "#4A1930", "#E8C282", "#E7DBBC", 
                     "#A68486", "#7900D7", "#A77500", "#6367A9", "#A05837", "#6B002C", "#772600", "#D790FF", "#9B9700",
                    "#549E79", "#FFF69F", "#201625", "#72418F", "#BC23FF", "#99ADC0", "#3A2465", "#922329",
                    "#5B4534", "#FDE8DC", "#404E55", "#0089A3", "#CB7E98", "#A4E804", "#324E72", "#6A3A4C",
                    "#83AB58", "#001C1E", "#D1F7CE", "#004B28", "#C8D0F6", "#A3A489", "#806C66", "#222800",
                    "#BF5650", "#E83000", "#66796D", "#DA007C", "#FF1A59", "#8ADBB4", "#1E0200", "#5B4E51",
                    "#C895C5", "#320033", "#FF6832", "#66E1D3", "#CFCDAC", "#D0AC94", "#7ED379", "#012C58",
                    "#7A7BFF", "#D68E01", "#353339", "#78AFA1", "#FEB2C6", "#75797C", "#837393", "#943A4D",
                    "#B5F4FF", "#D2DCD5", "#9556BD", "#6A714A", "#001325", "#02525F", "#0AA3F7", "#E98176",
                    "#DBD5DD", "#5EBCD1", "#3D4F44", "#7E6405", "#02684E", "#962B75", "#8D8546", "#9695C5",
                    "#E773CE", "#D86A78", "#3E89BE", "#CA834E", "#518A87", "#5B113C", "#55813B", "#E704C4",
                    "#00005F", "#A97399", "#4B8160", "#59738A", "#FF5DA7", "#F7C9BF", "#643127", "#513A01",
                    "#6B94AA", "#51A058", "#A45B02", "#1D1702", "#E20027", "#E7AB63", "#4C6001", "#9C6966",
                    "#64547B", "#97979E", "#006A66", "#391406", "#F4D749", "#0045D2", "#006C31", "#DDB6D0",
                    "#7C6571", "#9FB2A4", "#00D891", "#15A08A", "#BC65E9", "#FFFFFE", "#C6DC99", "#203B3C",
                    "#671190", "#6B3A64", "#F5E1FF", "#FFA0F2", "#CCAA35", "#374527", "#8BB400", "#797868",
                    "#C6005A", "#3B000A", "#C86240", "#29607C", "#402334", "#7D5A44", "#CCB87C", "#B88183",
                    "#AA5199", "#B5D6C3", "#A38469", "#9F94F0", "#A74571", "#B894A6", "#71BB8C", "#00B433",
                    "#789EC9", "#6D80BA", "#953F00", "#5EFF03", "#E4FFFC", "#1BE177", "#BCB1E5", "#76912F",
                    "#003109", "#0060CD", "#D20096", "#895563", "#29201D", "#5B3213", "#A76F42", "#89412E",
                    "#1A3A2A", "#494B5A", "#A88C85", "#F4ABAA", "#A3F3AB", "#00C6C8", "#EA8B66", "#958A9F",
                    "#BDC9D2", "#9FA064", "#BE4700", "#658188", "#83A485", "#453C23", "#47675D", "#3A3F00",
                    "#061203", "#DFFB71", "#868E7E", "#98D058", "#6C8F7D", "#D7BFC2", "#3C3E6E", "#D83D66",
                    "#2F5D9B", "#6C5E46", "#D25B88", "#5B656C", "#00B57F", "#545C46", "#866097", "#365D25",
                    "#252F99", "#00CCFF", "#674E60", "#FC009C", "#92896B", "#1E2324", "#DEC9B2", "#9D4948")
#n.clusters <- length(levels(dtr.clusters))
zsg.cluster.cols <- zsg.cluster.cols[c(1:n.clusters)]
names(zsg.cluster.cols) <- levels(zsg.meta$ZsG.Cluster)

zsg.cols <- c("#8745b5", "#13cf00")
names(zsg.cols) <- c("ZsGn", "ZsGp")

age.cols <- viridis(option="magma", n=5)
names(age.cols) <- c("Wk1", "Wk4", "Wk16", "32wk", "52wk")
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=6.75, fig.width=7.75}
zsg.graph <- ggplot(zsg.meta, aes(x=V1, y=V2)) +
  # geom_segment(data=zsg.edges, aes(x=from.x, xend=to.x, y=from.y, yend=to.y),
  #              colour='grey80', alpha=0.5) +
  geom_point_rast(aes(fill=ZsG.Cluster), shape=21, size=1.5) +
  theme_classic() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_Publication() +
  theme(axis.title = element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
	      axis.line=element_blank(),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  #guides(fill=FALSE) +
  NULL

ggsave(zsg.graph,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ZsG_graph-clusters.pdf",
       useDingbats=FALSE, height=6.75, width=7.75)

zsg.graph
```

This is the SNN-graph of entire experiment. I've omitted the edges deliberately to save on plotting time. What are these clusters?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=7.75, fig.width=14.75}
zsg.graph <- ggplot(zsg.meta, aes(x=V1, y=V2)) +
  # geom_segment(data=zsg.edges, aes(x=from.x, xend=to.x, y=from.y, yend=to.y),
  #              colour='grey80', alpha=0.5) +
  geom_point_rast(aes(fill=ZsG.Cluster), shape=21, size=2) +
  theme_classic() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_Publication() +
  theme(axis.title = element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
	      axis.line=element_blank(),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(title="Cluster", ncol=1, title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  facet_grid(SortType~Age) +
  NULL

ggsave(zsg.graph,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ZsG_graph-Age_by_ZsG.pdf",
       useDingbats=FALSE, height=7.75, width=14.75)

zsg.graph
```

It's pretty hard to see if there are any changes in density in these graphs based on these plots, though it does look like there might be _some_ differences. Can we 
get an idea of the expression changes by overlaying gene expression information? I'll need the SCE object (if it fits in memory) and `schex` for this.

__NB__ I can't fit the SCE object in memory so I need to make these plots on the clusters. I've run this with a resolution of 150 bins as this seems to generally give 
pretty reasonable resolution for visualisations with this number of cells.

Instead I'll plot the cluster proportions. I've already calculated the proportions for each cluster so I don't need to re-calculate these.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
zsg.props <- read.table("~/CI_filesystem/mnt/scratchb/jmlab/morgan02/10X/Thymus_ZsG/ZsG_cluster_props.tsv",
                        sep="\t", header=TRUE, stringsAsFactors=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=8.75, fig.width=10.75}
zsg.props$Age <- factor(zsg.props$Age,
                        levels=c("Wk1", "Wk4", "Wk16"),
                        labels=c("Wk1", "Wk4", "Wk16"))

label.box <- ggplot(zsg.props,
       aes(x=Age, y=value)) +
  geom_boxplot(fill='grey80') +
  scale_fill_manual(values=zsg.cols) +
  #scale_fill_Publication() +
  theme_mike() +
  labs(x="Time point (hours)", y="Proportion") +
  facet_wrap(~ZsG.Cluster, scales="free") +
  expand_limits(y=c(0))

ggsave(label.box,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ZsG_Cluster_boxplot.pdf",
       height=8.75, width=10.75, useDingbats=FALSE)

label.box 
```

There are fluctuations in the different clusters, but there is also quite a lot of variance in them too. Hopefully splitting them by ZsG-sorting population will resolve this 
heterogeneity.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=8.75, fig.width=10.75}
label.box <- ggplot(zsg.props,
       aes(x=Age, y=value, fill=SortType)) +
  geom_boxplot() +
  scale_fill_manual(values=zsg.cols) +
  #scale_fill_Publication() +
  theme_mike() +
  labs(x="Time point (hours)", y="Proportion") +
  facet_wrap(~ZsG.Cluster, scales="free") +
  expand_limits(y=c(0))

ggsave(label.box,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ZsG_Cluster-ZsGstate_boxplot.pdf",
       height=8.75, width=10.75, useDingbats=FALSE)

label.box 
```

For many clusters we can see that there is some clear separation between the ZsG+ and ZsG- proportions - but not all of them.

Can I use plots of marker genes to roughly assign labels to clusters?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Cd52)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

_Cd52_ is quite high in many clusters - the enrichment appears to be in clusters 1, 3, 6, 14, 16, 20 and 22. Only a subset of these actually express _Aire_.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Aire)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Namely clusters 1, 3 and 12. Interestingly cluster 12 doesnot have very high expression of _Cd52_ - these might be proliferating TEC?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Cdk1)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Yes - cluster 12 are proliferating mTEC. The other dividing clusters are 2, 8, 9 and 24.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Top2a)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

_Top2a_ marks S-phase.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Ube2c)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

_Ube2c_ marks G2/M.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Dll4)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Expression of _Dll4_ is restricted to cTEC - clusters 9 and 13 are cTEC of some flavour.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Prss16)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Clusters 9 and 13 have the highest expression of _Prss16_ too which confirms them as cTEC - it also looks cluster 10 might be cortical. Are any of these potentiall 
perinatal cTEC?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Psmb11)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Gper1)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Hmm, there might be some perinatal cTEC mixed in with the mature cTEC? The perinatal express _Cxcl12_ but lack _Krt5_.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Cxcl12)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

_Cxcl12_ is highly expressed in both cTEC and intertypical TEC. If clusters 10 and 13 are cTEC, then does that mean clusters 4, 10 and 11 are potential intertypical TEC?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Ccl21a)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

_Ccl21a_ is the marker for intertypical TEC, but is also widely expressed in mTEC. However, the co-expression of _Cxcl12_ and _Ccl21a_ should mark them. That indicates that 
clusters 4, 7 and 10 are the best candidates for intertypical TECs.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Krt5)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

_Krt5_ is the other marker for intertypical TEC - the high expression in cluster 7 indicates they are also probably intertypical TEC. Cluster 9 look like it might contain 
perinatal cTEC but it's almost impossible to tell them apart from the mature mTEC.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Avil)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

_Avil_ quite specifically marks Tuft-like mTEC.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Cd177)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

_Cd177_ is quite specific to eTEC1 - cluster 23 seems to fit the bill. I can confirm this with _Car8_.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Car8)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Yep - cluster 23 it is.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Krt20)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Cluster 14 have quite specifically high _Krt20_ expression - what are these?!

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Krt10)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Cluster 17 have high _Krt10_ expression - corneocytes?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Spink5)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Hmm, _Spink5_ seems to mark post-Aire mTECs and corneocytes - that probably confirms cluster 17 as corneocyte-like. Does that also mean that we have potentially mis-labelled 
the Post-Aire mTEC? I'll label clusters 17 and 19 as post-Aire as an umbrella term. It might also resolve why there are 2 clusters of post-Aire cells in the original 
ageing experiment.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Pdpn)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Cluster 11 might also be junction/intertypical TEC as they expression _Pdpn_, _Prss16_ and _Krt5_.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Ly6a)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```

Cluster 5 have high exprssin of _Ly6a/Sca1_ - what does this mark them as?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.65, fig.width=7.25}
ggplot(zsg.meta, aes(x=ZsG.Cluster, y=Apoe)) +
  geom_boxplot(fill='grey80') +
  #scale_fill_manual(values=zsg.cols) +
  theme_mike()
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
zsg.meta$CellType <- "Unknown"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(4, 7, 10)] <- "Intertypical TEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(6, 14, 16, 20, 22)] <- "Mature mTEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(1, 3)] <- "Aire+ mTEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(2, 12, 24)] <- "Proliferating TEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(13)] <- "Mature cTEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(9)] <- "Perinatal cTEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(4, 7, 10, 11)] <- "Intertypical TEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(15, 18)] <- "Tuft-like mTEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(23)] <- "eTEC1"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(17, 19)] <- "Post-Aire mTEC"
zsg.meta$CellType[zsg.meta$ZsG.Cluster %in% c(5)] <- "Sca1+"

mapping.mat <- table(zsg.meta$ZsG.Cluster, zsg.meta$CellType) > 0

pheatmap(mapping.mat + 0, cluster_cols=FALSE, cluster_rows=FALSE, 
         filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cluster-celltype_mapping.pdf")

pheatmap(mapping.mat + 0, cluster_cols=FALSE, cluster_rows=FALSE)
```

At the moment Clusters 8 and 21 are the only one I can't figure out. Does this categorization give the expected changes in subtype proportions? Cluster 8 are proliferating, 
but that doesn't tell me what they are. They look like some other type of cTEC based on their position in the SNN-graph.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
cluster.table <- as.data.frame(xtabs(~ CellType + Age + SortType + CellOrigin, data=zsg.meta))

# cast into square matrix format to make normalisation easier
zsg.clust.mat <- dcast(cluster.table, CellType + SortType ~ Age + CellOrigin, value.var='Freq')

# normalise by column sums to make them comparable across samples and clusters
zsg.norm.mat <- as.data.frame(apply(zsg.clust.mat[, -c(1, 2)], 2, FUN=function(Q) Q/sum(Q)))
zsg.norm.mat$CellType <- zsg.clust.mat$CellType
zsg.norm.mat$SortType <- zsg.clust.mat$SortType

zsg.cluster.df <- melt(zsg.norm.mat, id.vars=c("CellType", "SortType"))
zsg.cluster.df$Age <- unlist(lapply(strsplit(as.character(zsg.cluster.df$variable), split="_", fixed=TRUE), FUN=function(K) paste0(K[1])))
zsg.cluster.df$Age <- factor(zsg.cluster.df$Age,
                             levels=c("Wk1", "Wk4", "Wk16"),
                             labels=c("Wk1", "Wk4", "Wk16"))
zsg.cluster.df$SampID <- unlist(lapply(strsplit(as.character(zsg.cluster.df$variable), split="_", fixed=TRUE), FUN=function(K) paste0(K[2])))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=10.75}
label.box <- ggplot(zsg.cluster.df,
       aes(x=Age, y=value)) +
  geom_boxplot(fill='grey80') +
  scale_fill_manual(values=zsg.cols) +
  #scale_fill_Publication() +
  theme_mike() +
  labs(x="Time point (hours)", y="Proportion") +
  facet_wrap(~CellType, scales="free", ncol=5) +
  expand_limits(y=c(0))

ggsave(label.box,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Composition_Subtype-boxplot.pdf",
       height=4.75, width=10.75, useDingbats=FALSE)

label.box 
```

There are fluctuations in the different clusters, but the proportions do seem a little out. The loss of Aire+ mTEC and proliferating TEC makes sense. The loss of cTECs 
might be due to the decline of perinatal cTECs.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=6.75, fig.width=10.75}
label.box <- ggplot(zsg.cluster.df,
       aes(x=Age, y=value, fill=SortType)) +
  geom_boxplot() +
  scale_fill_manual(values=zsg.cols) +
  #scale_fill_Publication() +
  theme_mike() +
  labs(x="Time point (hours)", y="Proportion") +
  facet_wrap(~CellType, scales="free", ncol=5) +
  expand_limits(y=c(0))

ggsave(label.box,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Composition_Subtype_ZsG-boxplot.pdf",
       height=4.75, width=10.75, useDingbats=FALSE)

label.box
```

Splitting by ZsG status again resolves this heterogeneity. Notably we can see the decline of ZsG- cTECs (probably perinatal cTEC), and the constant % of ZsG+ cTEC (again 
probably mature cTEC). The loss of Aire+ mTEC seems fairly constant, though highly variable, whilst the Mature mTEC fluctuate quite a lot but inconsistently. The post-Aire 
mTEC don't change much by age or ZsG status, whilst the proliferting TEC decline, again regardless of ZsG status. The Sca1+ cells are intresting, they definitely look like 
they are declining throughout.

The eTEC1 cluster is fairly constant over time which fits with the previous observations too. Likewise the estimates proportion of Intertypical TEC also increases with age, 
but quite specifically only in the ZsG+ fraction; the ZsG- fraction declines!

Can we model these counts to find statistical evidence of changes? I don't entirely trust either the clustering or my grouping of cells into these TEC subtypes - they are 
too noisy. How about using `Cydar` to model the different regions of the high-dimensional manifold as hyperspheres, then find which regions of the TEC landscape change 
over age, and between ZsG sorting groups.

I need to pick the number of 'features' in this case this refers to the number of PCs and the number of hyperspheres to sample the whole space.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# I should use the same number of PCs as I did for the SNN-graph and clustering to make them concordant.
zsg.pcs <- read.table("~/CI_filesystem/mnt/scratchb/jmlab/morgan02/10X/Thymus_ZsG/ZsG_PCs.tsv",
                      sep="\t", header=TRUE, stringsAsFactor=FALSE)
rownames(zsg.pcs) <- zsg.pcs$Sample
n.dim <- 20
zsg.pcs.meta <- merge(zsg.meta, zsg.pcs, by='Sample')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# treat each age as a separate sample/condition
thymus.red <- zsg.pcs.meta[, c("SortType", "ZsG.Cluster", "Sample", "CellOrigin", "Age", paste0("PC", 1:n.dim))]

# need replicates for DA testing
thymus.list <- list()
for(x in seq_along(unique(zsg.pcs.meta$CellOrigin))){
  plate <- unique(zsg.pcs.meta$CellOrigin)[x]
  plate.red <- thymus.red[thymus.red$CellOrigin == plate, ]
  plate.ages <- unique(plate.red$Age)
  for(i in seq_along(plate.ages)){
    age <- unique(plate.ages)[i]
    age.red <- plate.red[plate.red$Age == age, ]
    age.zsg <- unique(age.red$SortType)
    for(k in seq_along(age.zsg)){
      stype <- age.zsg[k]
      zsg.red <- age.red[age.red$SortType == stype, ]
      
      age.mat <- as(zsg.red[, paste0("PC", 1:n.dim)], "matrix")
      rownames(age.mat) <- zsg.red$Sample
      thymus.list[[paste(age, paste0(plate), stype, sep=".")]] <- age.mat
    }
  }
}

age.condition <- unlist(lapply(strsplit(names(thymus.list),
                                        split=".", fixed=TRUE),
                               FUN=function(X) paste0(X[1])))
sort.condition <- unlist(lapply(strsplit(names(thymus.list),
                                        split=".", fixed=TRUE),
                               FUN=function(X) paste0(X[3])))

cydar.meta <- data.frame("Sample"=names(thymus.list), "Age"=age.condition,
                         "SortType"=sort.condition)
cydar.meta$AgeInt <- as.numeric(gsub(cydar.meta$Age, pattern="Wk", replacement=""))
cydar.meta$Age <- factor(cydar.meta$Age,
                         levels=c("Wk1", "Wk4", "Wk16"),
                         labels=c("Wk1", "Wk4", "Wk16"))

cydar.meta$Combo <- paste(cydar.meta$Age, cydar.meta$SortType, sep=".")

thymus.cydar <- prepareCellData(thymus.list)
```



```{r, echo=FALSE, warning=FALSE, message=FALSE}
thymus.dist <- neighborDistances(thymus.cydar, neighbors=31, as.tol=TRUE)
boxplot(thymus.dist)
```

This box plot shows the distribution of Euclidean distances in the 20-PC space for NNs from 1 to 20. These distances are expressed as tolerance values which is used to 
count cells in hyperspheres. This is the tolerance value that is proportional to the hypersphere radius, and so is quite important. It looks like that at around 15-20 
neighbours this starts to plateau. The median tolerance at 15 neighbours is 0.901, whilst at 20 neighbours it is 0.922. I'll go with 0.9 based on this.

I'll start by using age as the predictor variable.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# the consensus indicates the cut off should be 0.9
# Cydar defines the initial hyperspheres by uniformly sampling every Nth cell, where N=downsample - this 
# means that some regions in high-dimensional space get sampled more than others that is proportional to 
# their density.

thymus.cydar <- countCells(thymus.cydar, tol=0.9, filter=0, downsample=3)
message(paste0("Created ", nrow(thymus.cydar), " hyperspheres"))

# do DA testing with edgeR
thymus.dge <- DGEList(assay(thymus.cydar), lib.size=thymus.cydar$totals)

# filter low abundance hyperspheres
keep <- aveLogCPM(thymus.dge) >= aveLogCPM(2, mean(thymus.cydar$totals))
thymus.cydar <- thymus.cydar[keep,]
thymus.dge <- thymus.dge[keep,]

thymus.design <- model.matrix(~AgeInt + SortType, data=cydar.meta)
thymus.dge <- estimateDisp(thymus.dge, thymus.design)
# colnames(thymus.design) <- gsub(colnames(thymus.design), pattern=":", replacement="_")
# cydar.contrast <- makeContrasts(AgeInt_SortTypeZsGn - AgeInt_SortTypeZsGp,
#                                 levels=thymus.design)

thymus.fit <- glmQLFit(thymus.dge, thymus.design, robust=TRUE)
thymus.res <- glmQLFTest(thymus.fit, coef=2)
# thymus.res <- glmQLFTest(thymus.fit, contrast = cydar.contrast)

# control the spatial FDR
qvals <- spatialFDR(intensities(thymus.cydar), thymus.res$table$PValue)
is.sig <- qvals <= 0.01
summary(is.sig)
```

There are 2088 DA hyperspheres across the interaction between age and ZsG groups.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
sig.coords <- intensities(thymus.cydar)#[is.sig,]
sig.res <- thymus.res$table#[is.sig,]
coords <- prcomp(sig.coords)

plotCellLogFC(coords$x[,1], coords$x[,2], sig.res$logFC, xlab="Hypersphere PC1", ylab="Hypersphere PC2")
```

Each of these plots represents a hypersphere. The red shows increasing abundance with age, whilst the blue shows decreasing. This is an oversimplification as this only shows 
the first 2 PCs of all of hyperspheres (non-DA spheres are in grey). I could create a UMAP of the hyperspheres, or even an SNN-graph for visualisation.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
par(mfrow=c(3,3), mar=c(2.1, 1.1, 3.1, 1.1))
limits <- intensityRanges(thymus.cydar, p=0.01)
all.markers <- markernames(thymus.cydar)[1:9]
for (i in order(all.markers)) { 
  plotSphereIntensity(coords$x[,1], coords$x[,2], sig.coords[,i],
                    irange=limits[,i], main=all.markers[i],
                    cex=1.5)
}

```

This plot shows the median single-cell PC value for each hypersphere for the first 9 PCs overlaid on a separate PCA of the hyperspheres themselves. It mostly shows how 
different PCs define the different directions in the single-cell data.

I should be able to translate these hyperspheres back into regions in my reduced dimensional graph. How about taking the majority vote of cells in each hypersphere as the 
cluster ID?

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
cydar.umap <- umap(coords$x, 
                   n_neighbors=19,
                   init='random',
                   metric='euclidean',
                   n_components=2,
                   min_dist=0.4,
                   method='naive')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.95, fig.width=5.95}
cydar.umap.df <- data.frame("HS"=paste0("HS", c(1:nrow(thymus.cydar))), "UMAP.1"=cydar.umap$layout[, 1], "UMAP.2"=cydar.umap$layout[, 2],
                            "logFC"=sig.res$logFC)
# set the non-sig logFC to 0
cydar.umap.df$logFC[!is.sig] <- 0

ggplot(cydar.umap.df, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.umap.df[, c("UMAP.1", "UMAP.2")],
             shape=21, size=2.5, fill='grey80') +
  geom_point(data=cydar.umap.df[abs(cydar.umap.df$logFC) > 0, ],
             aes(fill=logFC), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_gradient2(low="blue", mid='grey80', high="red") +
  NULL

ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-ageFC.pdf",
       height=5.95, width=5.95, useDingbats=FALSE)
```

This is a UMAP representation of the hyperspheres. I want to show how different regions either increase or decrease in abundance. Areas in blue decrease in abundance with 
age, whilst the red ones increase. What are these regions though? Can I assign some meta-data to them, such as cluster and ZsG information?

I can extract the indices of the cells that fall into each hypersphere. Using this I can then plot the average age, major cluster and sort type in each hypersphere, etc.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
cell.ids <- colnames(cellIntensities(thymus.cydar))
cell.indices <- thymus.cydar@int_elementMetadata$cydar@listData$cellAssignments
# loop over the hyperspheres and calculate the proportion of cells in each cluster, proportion of cells at each age, as well as 
# the average expression of marker genes and proportion of ZsG+ cells

zsg.pcs.meta$SortType <- factor(zsg.pcs.meta$SortType)

hs.meta.list <- list()
for(x in seq_along(1:length(cell.indices))){
 x.cells <- cell.indices[[x]]
 x.cellids <- cell.ids[x.cells]
 x.ages <- gsub(as.character(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Age), pattern="Wk", replacement="")
 # use the median age of cells
 age.median <- median(as.numeric(x.ages))
 
 x.sort <- zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$SortType
 sort.tab <- as.data.frame(table(x.sort)/length(x.sort))
 
 x.cluster <- zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$ZsG.Cluster
 clust.tab <- as.data.frame(table(x.cluster)/length(x.cluster))
 
 # select the majority cluster too
 maj.clust <- as.character(clust.tab$x.cluster[clust.tab$Freq == max(clust.tab$Freq)])
 if(length(maj.clust) > 1){
   maj.clust <- paste(maj.clust, collapse=".")
 }
 
 x.meta <- do.call(cbind.data.frame,
                   list("clust"=dcast(clust.tab, . ~x.cluster, value.var='Freq')[, -1],
                        "sort"=dcast(sort.tab, . ~x.sort, value.var='Freq')[, -1]))
 x.meta$Median.Age <- age.median
 x.meta$Maj.Cluster <- maj.clust
 
 # now I need to summarise each marker gene
 x.cdk1 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Cdk1)
 x.meta$Cdk1 <- x.cdk1
 
 x.aire <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Aire)
 x.meta$Aire <- x.aire
 
 x.cd52 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Cd52)
 x.meta$Cd52 <- x.cd52
 
 x.prss16 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Prss16)
 x.meta$Prss16 <- x.prss16
 
 x.psmb11 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Psmb11)
 x.meta$Psmb11 <- x.psmb11
 
 x.avil <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Avil)
 x.meta$Avil <- x.avil
 
 x.dll4 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Dll4)
 x.meta$Dll4 <- x.dll4
 
 x.ccl21a <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Ccl21a)
 x.meta$Ccl21a <- x.ccl21a
 
 x.cxcl12 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Cxcl12)
 x.meta$Cxcl12 <- x.cxcl12
 
 x.spink5 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Spink5)
 x.meta$Spink5 <- x.spink5
 
 x.gper1 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Gper1)
 x.meta$Gper1 <- x.gper1
 
 x.cd177 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Cd177)
 x.meta$Cd177 <- x.cd177
 
 x.pdpn <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Pdpn)
 x.meta$Pdpn <- x.pdpn
 
 x.car8 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Car8)
 x.meta$Car8 <- x.car8
 
 x.krt5 <- mean(zsg.pcs.meta[zsg.pcs.meta$Sample %in% x.cellids, ]$Krt5)
 x.meta$Krt5 <- x.krt5
 
 hs.meta.list[[paste0(x)]] <- x.meta
}

hs.meta <- do.call(rbind.data.frame,
                   hs.meta.list)
hs.meta$HS <- paste0("HS", 1:nrow(hs.meta))
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.95, fig.width=5.95}
cydar.umap.merge <- merge(hs.meta, cydar.umap.df, by='HS')
# cydar.umap.merge$Median.Age <- factor(cydar.umap.merge$Median.Age,
#                                       levels=c(1, 4, 16),
#                                       labels=c(1, 4, 16))

ggplot(cydar.umap.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(aes(fill=Median.Age), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_distiller(palette="OrRd", direction=1) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-MedianAge.pdf",
       height=5.95, width=5.95, useDingbats=FALSE)
```

This shows the median age in each hypersphere. We can do the same analysis but comparing ZsG sorting.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.95, fig.width=5.95}
ggplot(cydar.umap.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(aes(fill=sort.ZsGp), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_gradient(low='#c63af3', high='#12f000', limits=c(0, 1)) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-ZsGProp.pdf",
       height=5.95, width=5.95, useDingbats=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
thymus.res.sort <- glmQLFTest(thymus.fit, coef=3)
# thymus.res <- glmQLFTest(thymus.fit, contrast = cydar.contrast)

# control the spatial FDR
qvals <- spatialFDR(intensities(thymus.cydar), thymus.res.sort$table$PValue)
sort.is.sig <- qvals <= 0.01
summary(sort.is.sig)
```

There are 1916 DA hyperspheres w.r.t. ZsG-sorting group. How many of these are also DA w.r.t. age?

```{r, echo=FALSE, warning=FALSE, message=FALSE}
table(sort.is.sig, is.sig)
```

There are 697 hyperspheres that are DA for both age and sorting type - these are probably the ones of greatest interest.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.95, fig.width=5.95}
cydar.sortlf.df <- data.frame("HS"=paste0("HS", c(1:nrow(thymus.cydar))), 
                              "logFC.Sort"=thymus.res.sort$table$logFC)
# set the non-sig logFC to 0
cydar.sortlf.df$logFC.Sort[!sort.is.sig] <- 0

cydar.merge.res <- merge(cydar.umap.merge, cydar.sortlf.df, by='HS')


ggplot(cydar.merge.res, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.res[, c("UMAP.1", "UMAP.2")],
             shape=21, size=2.5, fill='grey80') +
  geom_point(data=cydar.merge.res[abs(cydar.merge.res$logFC.Sort) > 0, ],
             aes(fill=logFC.Sort), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_gradient2(low="blue", mid='grey80', high="red") +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-sortFC.pdf",
       height=5.95, width=5.95, useDingbats=FALSE)
```

I'll now test for the interaction between age and sorting.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
thymus.design <- model.matrix(~AgeInt:SortType, data=cydar.meta)
thymus.dge <- estimateDisp(thymus.dge, thymus.design)
colnames(thymus.design) <- gsub(colnames(thymus.design), pattern=":", replacement="_")
cydar.contrast <- makeContrasts(AgeInt_SortTypeZsGp - AgeInt_SortTypeZsGn,
                                levels=thymus.design)

thymus.fit <- glmQLFit(thymus.dge, thymus.design, robust=TRUE)
#thymus.res <- glmQLFTest(thymus.fit, coef=2)
thymus.res.int <- glmQLFTest(thymus.fit, contrast = cydar.contrast)

# control the spatial FDR
qvals <- spatialFDR(intensities(thymus.cydar), thymus.res.int$table$PValue)
is.sig.int <- qvals <= 0.01
summary(is.sig.int)
```

There are 1270 hyperspheres for which there is a significant interaction between age and ZsG-sorting.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.95, fig.width=5.95}
cydar.interact.df <- data.frame("HS"=paste0("HS", c(1:nrow(thymus.cydar))), 
                                "logFC.Int"=thymus.res.int$table$logFC)
# set the non-sig logFC to 0
cydar.interact.df$logFC.Int[!is.sig.int] <- 0

cydar.merge.int <- merge(cydar.merge.res, cydar.interact.df, by='HS')


ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[, c("UMAP.1", "UMAP.2")],
             shape=21, size=2.5, fill='grey80') +
  geom_point(data=cydar.merge.int[abs(cydar.merge.int$logFC.Int) > 0, ],
             aes(fill=logFC.Int), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_gradient2(low="blue", mid='grey80', high="red") +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-InteractFC.pdf",
       height=5.95, width=5.95, useDingbats=FALSE)
```

This shows the log fold change for the interaction term between age and ZsG sorting. Positive LFchange is higher in the ZsG+.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
# Assign outlier cells to a specific age
cydar.merge.int$Age <- factor(cydar.merge.int$Median.Age,
                              levels=c(1, 4, 16),
                              labels=c(1, 4, 16))
cydar.merge.int$Age[is.na(cydar.merge.int$Age) & cydar.merge.int$Median.Age == 2.5] <- 4
cydar.merge.int$Age[is.na(cydar.merge.int$Age) & cydar.merge.int$Median.Age == 10] <- 16

ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=sort.ZsGp), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_gradient(low='#c63af3', high='#12f000', limits=c(0, 1)) +
  facet_wrap(~Age) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-ZsGPropby_age.pdf",
       height=4.95, width=9.95, useDingbats=FALSE)
```

What do these DA regions correspond to in terms of TEC subtypes?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Aire), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_viridis(limits=c(0, 5)) +
  facet_wrap(~Age) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-Aire_byAge.pdf",
       height=4.95, width=9.95, useDingbats=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Cd52), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_viridis(limits=c(0, 5)) +
  facet_wrap(~Age) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-Cd52_byAge.pdf",
       height=4.95, width=9.95, useDingbats=FALSE)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Ccl21a), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_viridis(limits=c(0, 9)) +
  facet_wrap(~Age) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-Ccl21a_byAge.pdf",
       height=4.95, width=9.95, useDingbats=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Krt5), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_viridis(limits=c(0, 6)) +
  facet_wrap(~Age) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-Krt5_byAge.pdf",
       height=4.95, width=9.95, useDingbats=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Pdpn), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_viridis(limits=c(0, 4)) +
  facet_wrap(~Age) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-Pdpn_byAge.pdf",
       height=4.95, width=9.95, useDingbats=FALSE)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Avil), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_viridis(limits=c(0, 5)) +
  facet_wrap(~Age) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-Avil_byAge.pdf",
       height=4.95, width=9.95, useDingbats=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=9.95, fig.width=9.95}
ggplot(cydar.merge.int, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=cydar.merge.int[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Median.Age), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_distiller(palette="OrRd", direction=1) +
  facet_wrap(~Maj.Cluster, ncol=4) +
  NULL
 
ggsave(filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cydar_UMAP-MajCluster.pdf",
       height=9.95, width=9.95, useDingbats=FALSE)
```

There's complexity in these data that might be resolved by a diffusion map - I'll use the PCs as the input data.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
rownames(zsg.pcs.meta) <- zsg.pcs.meta$Sample
zsg.dm <- DiffusionMap(as.matrix(zsg.pcs.meta[, paste0("PC", 1:20)]),
                           n_eigs=20, k=20, distance='euclidean')
plot(zsg.dm)
```

Ooh that looks pretty! I'd say there are ~3 lineages present in there. __NB__ This is insanely slow.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
zsg.dpt <- DPT(zsg.dm)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
zsg.diffusion <- do.call(cbind.data.frame,
                         list("DC1"=zsg.dpt$DC1, "DC2"=zsg.dpt$DC2, "DC3"=zsg.dpt$DC3,
                             "DC4"=zsg.dpt$DC4, "DC5"=zsg.dpt$DC4, "DC6"=zsg.dpt$DC4,
                             "DC7"=zsg.dpt$DC4, "DC8"=zsg.dpt$DC4, "DC9"=zsg.dpt$DC4,
                             "DC10"=zsg.dpt$DC4, "DPT1"=zsg.dpt$DPT1, "DPT2"=zsg.dpt$DPT2,
                             "DPT3"=zsg.dpt$DPT3, "Branch"=zsg.dpt$Branch, "Sample"=zsg.pcs.meta$Sample))

rownames(zsg.diffusion) <- zsg.diffusion$Sample
zsg.diffusion.merge <- merge(zsg.diffusion, zsg.pcs.meta, by='Sample')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(zsg.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=zsg.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=SortType), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=zsg.cols) +
  facet_wrap(~Age, ncol=4) +
  NULL
```

It looks like the 3 lineages all separate on different diffusion components.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(zsg.diffusion.merge, aes(x=DC2, y=DC3)) +
  geom_point(data=zsg.diffusion.merge[,c("DC2", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=SortType), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=zsg.cols) +
  facet_wrap(~Age, ncol=4) +
  NULL
```

I need to determine which cell is the apex of these branching lineages.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(zsg.diffusion.merge, aes(x=DC3, y=DC4)) +
  geom_point(data=zsg.diffusion.merge[,c("DC3", "DC4")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=SortType), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=zsg.cols) +
  facet_wrap(~Age, ncol=4) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.95, fig.width=9.95}
ggplot(zsg.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=zsg.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  facet_wrap(~Age, ncol=4) +
  NULL
```

There's a lot to unpack here - namely that DC1 appears to separate one of the mTEC lineages. It might just be 
easier to pull out the cells inthe clusters along each branch, including everything at the apex, and try to 
identify the lineages that way.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=9.95, fig.height=13.95}
ggplot(zsg.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=zsg.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  facet_wrap(~ZsG.Cluster, ncol=4) +
  NULL
```

Clusters 6 and 22 appear to form the later mTEC lineage. Clusters 8, 9, 11 and 21 appear to form the other unknown lineage trajectory.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=9.95, fig.height=13.95}
ggplot(zsg.diffusion.merge, aes(x=DC2, y=DC3)) +
  geom_point(data=zsg.diffusion.merge[,c("DC2", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  facet_wrap(~ZsG.Cluster, ncol=4) +
  NULL
```

The separation DC3 appears to the Tuft-like mTEC lineage. That would be clusters 15 and 18.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=9.95, fig.height=13.95}
ggplot(zsg.diffusion.merge, aes(x=DC3, y=DC4)) +
  geom_point(data=zsg.diffusion.merge[,c("DC3", "DC4")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  facet_wrap(~ZsG.Cluster, ncol=4) +
  NULL
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
branch1.cells <- zsg.pcs.meta$Sample[zsg.pcs.meta$ZsG.Cluster %in% c(8, 10, 11, 21)]
branch1.meta <- zsg.pcs.meta[branch1.cells, ]
rownames(branch1.meta) <- branch1.meta$Sample

branch1.dm <- DiffusionMap(as.matrix(branch1.meta[, paste0("PC", 1:20)]),
                           n_eigs=20, k=20, distance='euclidean')
plot(branch1.dm)
```

It looks like there might be one trajectory and then a separate population? It does look a bit like one of the clusters 
feeds into this other lineage though...

```{r, echo=FALSE, warning=FALSE, message=FALSE}
branch1.dpt <- DPT(branch1.dm)

branch1.diffusion <- do.call(cbind.data.frame,
                         list("DC1"=branch1.dpt$DC1, "DC2"=branch1.dpt$DC2, "DC3"=branch1.dpt$DC3,
                             "DC4"=branch1.dpt$DC4, "DC5"=branch1.dpt$DC4, "DC6"=branch1.dpt$DC4,
                             "DC7"=branch1.dpt$DC4, "DC8"=branch1.dpt$DC4, "DC9"=branch1.dpt$DC4,
                             "DC10"=branch1.dpt$DC4, "DPT1"=branch1.dpt$DPT1, "DPT2"=branch1.dpt$DPT2,
                             "DPT3"=branch1.dpt$DPT3, "Branch"=branch1.dpt$Branch,
                             "Sample"=branch1.meta$Sample))

rownames(branch1.diffusion) <- branch1.diffusion$Sample
branch1.diffusion.merge <- merge(branch1.diffusion, branch1.meta, by='Sample')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(branch1.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=branch1.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=SortType), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cols) +
  facet_wrap(~Age, ncol=4) +
  NULL
```

Is that two separate lineages?

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(branch1.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=branch1.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  facet_wrap(~Age, ncol=4) +
  NULL
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=9.75, fig.height=12.75}
ggplot(branch1.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=branch1.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  facet_grid(ZsG.Cluster~Age) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(branch1.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=branch1.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Cdk1), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  NULL
```

These look like completely different lineages, but some cells from cluster 11 look like they might be transitioning into 
cluster 21 or _vice versa_.

What about looking just the putative intertypical TEC?

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
itec.cells <- zsg.pcs.meta$Sample[zsg.pcs.meta$ZsG.Cluster %in% c(4, 7, 10, 11)]
itec.meta <- zsg.pcs.meta[itec.cells, ]
rownames(itec.meta) <- itec.meta$Sample

itec.dm <- DiffusionMap(as.matrix(itec.meta[, paste0("PC", 1:20)]),
                           n_eigs=20, k=20, distance='euclidean')
plot(itec.dm)
```

It looks like there might be one trajectory and then a separate population? It does look a bit like one of the clusters 
feeds into this other lineage though...

```{r, echo=FALSE, warning=FALSE, message=FALSE}
itec.dpt <- DPT(itec.dm)

itec.diffusion <- do.call(cbind.data.frame,
                         list("DC1"=itec.dpt$DC1, "DC2"=itec.dpt$DC2, "DC3"=itec.dpt$DC3,
                             "DC4"=itec.dpt$DC4, "DC5"=itec.dpt$DC4, "DC6"=itec.dpt$DC4,
                             "DC7"=itec.dpt$DC4, "DC8"=itec.dpt$DC4, "DC9"=itec.dpt$DC4,
                             "DC10"=itec.dpt$DC4, "DPT1"=itec.dpt$DPT1, "DPT2"=itec.dpt$DPT2,
                             "DPT3"=itec.dpt$DPT3, "Branch"=itec.dpt$Branch,
                             "Sample"=itec.meta$Sample))

rownames(itec.diffusion) <- itec.diffusion$Sample
itec.diffusion.merge <- merge(itec.diffusion, itec.meta, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(itec.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=itec.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  NULL
```

Hmm, not a lot of obvious structure in there...

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(itec.diffusion.merge, aes(x=DC2, y=DC3)) +
  geom_point(data=itec.diffusion.merge[,c("DC2", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  NULL
```

This make it look like cluster 7 is a lineage in a specific direction.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(itec.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=itec.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Psmb11), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  NULL
```

Hmm, there's a fair bit of _Psmb11_ expression in the cluster 10 cells - this either means that they are cTECs or TEPCs.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=9.75, fig.width=6.95}
ggplot(itec.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=itec.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Ccl21a), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_grid(ZsG.Cluster~Age) +
  NULL
```

Generally speaking, _Ccl21a_ is primarily expressed in clusters 4 and 7, though there is some overlap. It definitely looks like there is a subset of clusters 10 and 11 at all 3 time points. 
Is there a split with ZsG+/- status?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=9.75, fig.width=8.95}
ggplot(itec.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=itec.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Ccl21a), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_grid(ZsG.Cluster~SortType) +
  NULL
```

The separation between _Ccl21a_+/- isn't due to the lineage tracing populations either.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=9.75, fig.width=6.95}
ggplot(itec.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=itec.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Psmb11), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_grid(ZsG.Cluster~Age) +
  NULL
```

It looks, just by eye-balling this plot, that cluster 4 increases proportionally with age. I have a feeling that _Psmb11_ and _Ccl21a_ might not overlap in expression much.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.75, fig.width=9.95}
itec.diffusion.merge$Ccl21a.Psmb11 <- as.numeric(itec.diffusion.merge$Ccl21a > 0 & itec.diffusion.merge$Psmb11 > 0)

ggplot(itec.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=itec.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Ccl21a.Psmb11), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_grid(Age~ZsG.Cluster) +
  NULL
```

This shows which cells have both non-zero expression of _Psmb11_ and _Ccl21a_ by age and cluster. It looks like there is an overlap region which might be expected if the cells are 
differentiating. It would be nice to figure out if these cells are biased toward one lineage or another.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
tecdiff.cells <- zsg.pcs.meta$Sample[zsg.pcs.meta$ZsG.Cluster %in% c(1, 2, 3,  4, 7, 9, 10, 11, 12, 13, 14, 16)]
tecdiff.meta <- zsg.pcs.meta[tecdiff.cells, ]
rownames(tecdiff.meta) <- tecdiff.meta$Sample

tecdiff.dm <- DiffusionMap(as.matrix(tecdiff.meta[, paste0("PC", 1:20)]),
                           n_eigs=20, k=20, distance='euclidean')
plot(tecdiff.dm)
```

This includes the iTEC populations, as well as the putative cTEC and mTEC populations.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
tecdiff.dpt <- DPT(tecdiff.dm)

tecdiff.diffusion <- do.call(cbind.data.frame,
                         list("DC1"=tecdiff.dpt$DC1, "DC2"=tecdiff.dpt$DC2, "DC3"=tecdiff.dpt$DC3,
                             "DC4"=tecdiff.dpt$DC4, "DC5"=tecdiff.dpt$DC4, "DC6"=tecdiff.dpt$DC4,
                             "DC7"=tecdiff.dpt$DC4, "DC8"=tecdiff.dpt$DC4, "DC9"=tecdiff.dpt$DC4,
                             "DC10"=tecdiff.dpt$DC4, "DPT1"=tecdiff.dpt$DPT1, "DPT2"=tecdiff.dpt$DPT2,
                             "DPT3"=tecdiff.dpt$DPT3, "Branch"=tecdiff.dpt$Branch,
                             "Sample"=tecdiff.meta$Sample))

rownames(tecdiff.diffusion) <- tecdiff.diffusion$Sample
tecdiff.diffusion.merge <- merge(tecdiff.diffusion, tecdiff.meta, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(tecdiff.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=tecdiff.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL
```

It looks like there might be 3 lineages here, one separated on each of the first 3 diffusion components.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
ggplot(tecdiff.diffusion.merge, aes(x=DC2, y=DC3)) +
  geom_point(data=tecdiff.diffusion.merge[,c("DC2", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL
```

There are still too many trajectories. I'll remove the cTEC clusters 9 and 13.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
mtec.cells <- zsg.pcs.meta$Sample[zsg.pcs.meta$ZsG.Cluster %in% c(1, 3, 14)]
mtec.meta <- zsg.pcs.meta[mtec.cells, ]
rownames(mtec.meta) <- mtec.meta$Sample

mtec.dm <- DiffusionMap(as.matrix(mtec.meta[, paste0("PC", 1:20)]),
                           n_eigs=20, k=20, distance='euclidean')
plot(mtec.dm)
```

This is just a subset of the mTEC populations - that side trajectory looks pretty wild!

```{r, echo=FALSE, warning=FALSE, message=FALSE}
mtec.dpt <- DPT(mtec.dm)

mtec.diffusion <- do.call(cbind.data.frame,
                         list("DC1"=mtec.dpt$DC1, "DC2"=mtec.dpt$DC2, "DC3"=mtec.dpt$DC3,
                             "DC4"=mtec.dpt$DC4, "DC5"=mtec.dpt$DC4, "DC6"=mtec.dpt$DC4,
                             "DC7"=mtec.dpt$DC4, "DC8"=mtec.dpt$DC4, "DC9"=mtec.dpt$DC4,
                             "DC10"=mtec.dpt$DC4, "DPT1"=mtec.dpt$DPT1, "DPT2"=mtec.dpt$DPT2,
                             "DPT3"=mtec.dpt$DPT3, "Branch"=mtec.dpt$Branch,
                             "Sample"=mtec.meta$Sample))

rownames(mtec.diffusion) <- mtec.diffusion$Sample
mtec.diffusion.merge <- merge(mtec.diffusion, mtec.meta, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(mtec.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=mtec.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL
```

Clusters 1, 3, and 14 form a single bifurcating trajectory with the root cells in cluster 3.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
ggplot(mtec.diffusion.merge, aes(x=DC2, y=DC3)) +
  geom_point(data=mtec.diffusion.merge[,c("DC2", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
ggplot(mtec.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=mtec.diffusion.merge[,c("DC1", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
ggplot(mtec.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=mtec.diffusion.merge[,c("DC1", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Cd52), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colorbar(title="Cd52", title.theme=element_text(size=14),
                             override.aes=list(shape=22, size=3))) +
  NULL
```

Hmm, not all cells express _Cd52_, particularly those in tail end of cluster 1.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
ggplot(mtec.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=mtec.diffusion.merge[,c("DC1", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Aire), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colorbar(title="Aire", title.theme=element_text(size=14),
                             override.aes=list(shape=22, size=3))) +
  NULL
```

Those same cells don't express _Aire_ either. Are they post-Aire?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
ggplot(mtec.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=mtec.diffusion.merge[,c("DC1", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Spink5), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colorbar(title="Spink5", title.theme=element_text(size=14),
                             override.aes=list(shape=22, size=3))) +
  NULL
```

The tail end of cluster 3 express _Spink5_ indicating these are transitioning to a post-Aire state.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
ggplot(mtec.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=mtec.diffusion.merge[,c("DC1", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Ccl21a), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colorbar(title="Ccl21a", title.theme=element_text(size=14),
                             override.aes=list(shape=22, size=3))) +
  NULL
```

Interesting, the tail of cluster 1 express a fair amount of _Ccl21a_ - a marker of intertypical TEC.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
ggplot(mtec.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=mtec.diffusion.merge[,c("DC1", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Krt5), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colorbar(title="Krt5", title.theme=element_text(size=14),
                             override.aes=list(shape=22, size=3))) +
  NULL
```

Likewise with _Krt5_ - I think that tail of cluster 1 is where the iTEC -> mTEC transitioning cells are. I'll make a diffusion map using the iTEC clusters, and include cluster 1 as the mTEC.


```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
transition.cells <- zsg.pcs.meta$Sample[zsg.pcs.meta$ZsG.Cluster %in% c(1, 3, 4, 7, 12)]
transition.meta <- zsg.pcs.meta[transition.cells, ]
rownames(transition.meta) <- transition.meta$Sample

transition.dm <- DiffusionMap(as.matrix(transition.meta[, paste0("PC", 1:20)]),
                           n_eigs=20, k=21, distance='euclidean')
plot(transition.dm)
```

This is the iTEC combined with cluster 1 from the mTEC - it looks like there are 2 side populations, but they are pretty smooth - no big transition jumps.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
transition.dpt <- DPT(transition.dm)

transition.diffusion <- do.call(cbind.data.frame,
                         list("DC1"=transition.dpt$DC1, "DC2"=transition.dpt$DC2, "DC3"=transition.dpt$DC3,
                             "DC4"=transition.dpt$DC4, "DC5"=transition.dpt$DC4, "DC6"=transition.dpt$DC4,
                             "DC7"=transition.dpt$DC4, "DC8"=transition.dpt$DC4, "DC9"=transition.dpt$DC4,
                             "DC10"=transition.dpt$DC4, "DPT1"=transition.dpt$DPT1, "DPT2"=transition.dpt$DPT2,
                             "DPT3"=transition.dpt$DPT3, "Branch"=transition.dpt$Branch,
                             "Sample"=transition.meta$Sample))

rownames(transition.diffusion) <- transition.diffusion$Sample
transition.diffusion.merge <- merge(transition.diffusion, transition.meta, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(transition.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=transition.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL
```

This makes it look like some mTECs are derived from a lineage trajectory that passes through a transit-amplifying compartment, and a second trajectory that does not require any cell division.


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# add interpretations to the clusters
transition.diffusion.merge$TECtype <- "Unknown"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(1)] <- "mTEC.1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(3)] <- "mTEC.2"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(14)] <- "mTEC.3"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(6)] <- "mTEC.4"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(20)] <- "mTEC.5"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(22)] <- "mTEC.6"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(16)] <- "mTEC.7"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(4)] <- "Intertypical.TEC.1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(7)] <- "Intertypical.TEC.2"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(10)] <- "Intertypical.TEC.3"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(11)] <- "Intertypical.TEC.4"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(9)] <- "cTEC.1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(13)] <- "cTEC.2"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(17)] <- "PostAire.mTEC.1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(19)] <- "PostAire.mTEC.2"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(23)] <- "eTEC1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(24)] <- "Prolif.TEC.1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(12)] <- "Prolif.TEC.2"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(2)] <- "Prolif.TEC.3"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(15)] <- "Tuft.mTEC.1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(18)] <- "Tuft.mTEC.2"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(5)] <- "Sca1.TEC.1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(8)] <- "New.TEC.1"
transition.diffusion.merge$TECtype[transition.diffusion.merge$ZsG.Cluster %in% c(21)] <- "New.TEC.2"

# use variations on the original ageing cluster colour scheme
inter.cols <- c("#9970ab", "#a28aac", "#35978f", "#4dd7cc", "#2d8179", "#63b9b2",
                "#B0cdc1", "#99b2a7", "#762a83", "#561f60", "#01665e", "#1f625d", "#014944",
                "#e7d4e8", "#e9e5e9", "#d5c5d5", "#dfbfdf", "#c59cc5", "#dea6de", "#f8cdf8",
                "#dfc27d", "#0042ff", "#4d79f7", "#96a679")


names(inter.cols) <- c("PostAire.mTEC.1", "PostAire.mTEC.2", "Intertypical.TEC.1", "Intertypical.TEC.2", "Intertypical.TEC.3", "Intertypical.TEC.4",
                       'cTEC.1', 'cTEC.2', 'Tuft.mTEC.1', 'Tuft.mTEC.2', 'Prolif.TEC.1', 'Prolif.TEC.2', 'Prolif.TEC.3',
                       'mTEC.1', 'mTEC.2', 'mTEC.3', 'mTEC.4', 'mTEC.5', 'mTEC.6', 'mTEC.7',
                       'eTEC1', 'New.TEC.1', 'New.TEC.2', 'Sca1.TEC.1')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
transition.diffusion.merge$AgeFactor <- factor(transition.diffusion.merge$Age,
                                               levels=c("Wk1", "Wk4", "Wk16"),
                                               labels=c("Week 1", "Week 4", "Week 16"))

mtec.dm.p <- ggplot(transition.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point_rast(data=transition.diffusion.merge[,c("DC1", "DC2")], 
                  fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point_rast(aes(fill=TECtype), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  #scale_fill_viridis() +
  facet_wrap(~AgeFactor, ncol=4) +
  theme(strip.background=element_rect(fill='white', colour='white')) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  guides(fill=guide_legend(title="TEC.Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL


ggsave(mtec.dm.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/DiffusionMap_mTECiTEC_age.pdf",
       height=4.75, width=11.75, useDingbats=FALSE)

mtec.dm.p
```

This indicates that the decline of intertypical TEC.2 impacts on mTEC numbers, whilst the simultaneous decline proliferating TEC.2 also reduces the number of cells differentiating to maintain 
the mTEC compartment.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
transition.diffusion.merge$AgeFactor <- factor(transition.diffusion.merge$Age,
                                               levels=c("Wk1", "Wk4", "Wk16"),
                                               labels=c("Week 1", "Week 4", "Week 16"))

mtec.dm.p <- ggplot(transition.diffusion.merge, aes(x=DC2, y=DC3)) +
  geom_point_rast(data=transition.diffusion.merge[,c("DC2", "DC3")], 
                  fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point_rast(aes(fill=TECtype), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  #scale_fill_viridis() +
  facet_wrap(~AgeFactor, ncol=4) +
  theme(strip.background=element_rect(fill='white', colour='white')) +
  labs(x="Diffusion Component 2", y="Diffusion Component 3") +
  guides(fill=guide_legend(title="TEC.Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL


ggsave(mtec.dm.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/DiffusionMap_mTECiTEC-DC2DC3_age.pdf",
       height=4.75, width=11.75, useDingbats=FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=11.75}
transition.diffusion.merge$AgeFactor <- factor(transition.diffusion.merge$Age,
                                               levels=c("Wk1", "Wk4", "Wk16"),
                                               labels=c("Week 1", "Week 4", "Week 16"))

mtec.dm.p <- ggplot(transition.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=transition.diffusion.merge[,c("DC1", "DC3")], 
                  fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=TECtype), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  #scale_fill_viridis() +
  facet_wrap(~AgeFactor, ncol=4) +
  theme(strip.background=element_rect(fill='white', colour='white')) +
  labs(x="Diffusion Component 2", y="Diffusion Component 3") +
  guides(fill=guide_legend(title="TEC.Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL


mtec.dm.p
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=11.75, fig.width=9.75}
# select the cell index for the earlist iTEC
min.dc1 <- min(transition.diffusion.merge$DC1)
start.cell <- which(transition.diffusion.merge$DC1 == min.dc1)
transition.diffusion.merge$Trans.DPT <- transition.dpt[start.cell]

mtec.dpt.p <- ggplot(transition.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point_rast(data=transition.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point_rast(aes(fill=Trans.DPT), shape=21, size=2.5) +
  theme_mike() +
  scale_fill_viridis() +
  facet_grid(TECtype~AgeFactor) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  guides(fill=guide_colorbar(title="DPT", title.theme=element_text(size=14),
                             override.aes=list(shape=22, size=3))) +
  theme(strip.background=element_rect(fill='white', colour='white')) +
  NULL

ggsave(mtec.dpt.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ZsG_mTEC-DPT_raster.pdf",
       height=11.75, width=10.75, useDingbats=FALSE)

mtec.dpt.p
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=7.75, fig.width=5.25}
ggplot(transition.diffusion.merge, aes(x=Trans.DPT, fill=TECtype)) +
  geom_density() +
  facet_wrap(~AgeFactor, ncol=1) +
  scale_fill_manual(values=inter.cols) +
  theme_mike()
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(transition.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=transition.diffusion.merge[,c("DC1", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(transition.diffusion.merge, aes(x=DC2, y=DC3)) +
  geom_point(data=transition.diffusion.merge[,c("DC2", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=3))) +
  NULL
```

These diffusion maps make it look more like clusters 1 and 4 are continuous.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(transition.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=transition.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Cd52), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colourbar(title="Cd52", title.theme=element_text(size=14),
                              override.aes=list(shape=22, size=3))) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(transition.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=transition.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Psmb11), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colourbar(title="Psmb11", title.theme=element_text(size=14),
                              override.aes=list(shape=22, size=3))) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(transition.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=transition.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Ccl21a), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colourbar(title="Ccl21a", title.theme=element_text(size=14),
                              override.aes=list(shape=22, size=3))) +
  NULL
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(transition.diffusion.merge, aes(x=DC1, y=DC2)) +
  geom_point(data=transition.diffusion.merge[,c("DC1", "DC2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Prss16), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colourbar(title="Prss16", title.theme=element_text(size=14),
                              override.aes=list(shape=22, size=3))) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.75, fig.width=9.75}
ggplot(transition.diffusion.merge, aes(x=DC1, y=DC3)) +
  geom_point(data=transition.diffusion.merge[,c("DC1", "DC3")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=Cdk1), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_manual(values=zsg.cluster.cols) +
  scale_fill_viridis() +
  facet_wrap(~Age, ncol=4) +
  guides(fill=guide_colourbar(title="Cdk1", title.theme=element_text(size=14),
                              override.aes=list(shape=22, size=3))) +
  NULL
```

That little off-shoot of cluster 2 are dividing cells.

## UMAP

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(42)
tec.umap <-  umap(as.matrix(zsg.pcs.meta[, paste0("PC", 1:20)]),
                  n_components=2,
                  n_neighbors=31, metric='euclidean',
                  init='spectral', min_dist=0.3)

tec.umap.df <- as.data.frame(tec.umap$layout)
colnames(tec.umap.df) <- c("UMAP.1", "UMAP.2")
tec.umap.df$Sample <- zsg.pcs.meta$Sample

tec.umap.merge <- merge(zsg.pcs.meta, tec.umap.df, by='Sample')
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=7.75, fig.width=14.75}
tec.umap.merge$AgeFactor <- factor(tec.umap.merge$Age,
                                   levels=c("Wk1", "Wk4", "Wk16"),
                                   labels=c("Week 1", "Week 4", "Week 16"))

ggplot(tec.umap.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=tec.umap.merge[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~AgeFactor, ncol=4) +
  theme(strip.background=element_rect(fill='white', colour='white'),
        strip.text=element_text(size=16)) +
  guides(fill=guide_legend(title="TEC.Cluster", nrow=3, title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=4))) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=15.75, fig.width=14.75}
ggplot(tec.umap.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=tec.umap.merge[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.3, colour='grey80', shape=21) +
  geom_point(aes(fill=ZsG.Cluster), shape=21, size=1.8) +
  theme_mike() +
  scale_fill_manual(values=zsg.cluster.cols) +
  #scale_fill_viridis() +
  facet_wrap(~ZsG.Cluster, ncol=4) +
  guides(fill=guide_legend(title="Cluster", title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=4))) +
  NULL
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# add interpretations to the clusters
tec.umap.merge$TECtype <- "Unknown"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(1)] <- "mTEC.1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(3)] <- "mTEC.2"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(14)] <- "mTEC.3"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(6)] <- "mTEC.4"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(20)] <- "mTEC.5"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(22)] <- "mTEC.6"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(16)] <- "mTEC.7"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(4)] <- "Intertypical.TEC.1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(7)] <- "Intertypical.TEC.2"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(10)] <- "Intertypical.TEC.3"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(11)] <- "Intertypical.TEC.4"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(9)] <- "cTEC.1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(13)] <- "cTEC.2"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(17)] <- "PostAire.mTEC.1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(19)] <- "PostAire.mTEC.2"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(23)] <- "eTEC1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(24)] <- "Prolif.TEC.1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(12)] <- "Prolif.TEC.2"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(2)] <- "Prolif.TEC.3"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(15)] <- "Tuft.mTEC.1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(18)] <- "Tuft.mTEC.2"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(5)] <- "Sca1.TEC.1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(8)] <- "New.TEC.1"
tec.umap.merge$TECtype[tec.umap.merge$ZsG.Cluster %in% c(21)] <- "New.TEC.2"

# use variations on the original ageing cluster colour scheme
inter.cols <- c("#9970ab", "#a28aac", "#35978f", "#4dd7cc", "#2d8179", "#63b9b2",
                "#B0cdc1", "#99b2a7", "#762a83", "#561f60", "#01665e", "#1f625d", "#014944",
                "#e7d4e8", "#e9e5e9", "#d5c5d5", "#dfbfdf", "#c59cc5", "#dea6de", "#f8cdf8",
                "#dfc27d", "#0042ff", "#4d79f7", "#96a679")


names(inter.cols) <- c("PostAire.mTEC.1", "PostAire.mTEC.2", "Intertypical.TEC.1", "Intertypical.TEC.2", "Intertypical.TEC.3", "Intertypical.TEC.4",
                       'cTEC.1', 'cTEC.2', 'Tuft.mTEC.1', 'Tuft.mTEC.2', 'Prolif.TEC.1', 'Prolif.TEC.2', 'Prolif.TEC.3',
                       'mTEC.1', 'mTEC.2', 'mTEC.3', 'mTEC.4', 'mTEC.5', 'mTEC.6', 'mTEC.7',
                       'eTEC1', 'New.TEC.1', 'New.TEC.2', 'Sca1.TEC.1')
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=6.75, fig.width=11.75}
tec.umap.merge$AgeFactor <- factor(tec.umap.merge$Age,
                                   levels=c("Wk1", "Wk4", "Wk16"),
                                   labels=c("Week 1", "Week 4", "Week 16"))

ggplot(tec.umap.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point(data=tec.umap.merge[,c("UMAP.1", "UMAP.2")], 
             fill='grey80', size=1, alpha=0.5, shape=21) +
  geom_point(aes(fill=TECtype), shape=21, size=1.8) +
  theme_mike() +
  #scale_fill_Publication() +
  scale_fill_manual(values=inter.cols) +
  #scale_fill_viridis() +
  #facet_wrap(~AgeFactor, ncol=4) +
  theme(strip.background=element_rect(fill='white', colour='white'),
        strip.text=element_text(size=16), 
        legend.position="right") +
  guides(fill=guide_legend(title="TEC.Cluster", ncol=1, title.position="top",
                           title.theme=element_text(size=14),
                           override.aes=list(shape=22, size=4.5))) +
  NULL
```

Would it be better to have the cluster labels overlaid on the UMAP? I'll use the average dimensions for the cluster as the position for the label

```{r, echo=FALSE, warning=FALSE, message=FALSE}
umap1.mean <- unlist(by(data=tec.umap.merge$UMAP.1, INDICES=tec.umap.merge$TECtype, FUN=mean))
umap1.order <- names(umap1.mean)
umap2.mean <- unlist(by(data=tec.umap.merge$UMAP.2, INDICES=tec.umap.merge$TECtype, FUN=mean))
umap2.order <- names(umap2.mean)

umap.positions <- data.frame("Label"=umap1.order, "UMAP.1"=as.numeric(umap1.mean), "UMAP.2"=as.numeric(umap2.mean))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=6.75, fig.width=11.75}
umap.p <- ggplot(tec.umap.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point_rast(aes(color=TECtype), size=2.5) +
  geom_label_repel(data=umap.positions, 
                   aes(label=Label), fill='grey80',
                   size=4, colour='black') +
  theme_mike() +
  scale_color_manual(values=inter.cols) +
  theme(strip.background=element_rect(fill='white', colour='white'),
        strip.text=element_text(size=16), 
        legend.position="right") +
  guides(color=FALSE) +
  NULL

ggsave(umap.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/UMAP_raster_overview.pdf",
       height=6.75, width=7.75, useDingbats=TRUE)

umap.p
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.75, fig.width=15.75}
tec.umap.merge$AgeFactor <- factor(tec.umap.merge$Age,
                                   levels=c("Wk1", "Wk4", "Wk16"),
                                   labels=c("Week 1", "Week 4", "Week 16"))

age.umap.p <- ggplot(tec.umap.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point_rast(data=tec.umap.merge[,c("UMAP.1", "UMAP.2")], 
                  color='grey80', size=1, alpha=0.5) +
  geom_point_rast(aes(color=TECtype), size=2.5) +
  theme_mike() +
  scale_color_manual(values=inter.cols) +
  facet_wrap(~AgeFactor, ncol=4) +
  theme(strip.background=element_rect(fill='white', colour='white'),
        strip.text=element_text(size=21), 
        legend.position="right") +
  guides(color=guide_legend(title="TEC.Cluster", ncol=1, title.position="top",
                           title.theme=element_text(size=16),
                           override.aes=list(size=4))) +
  NULL

ggsave(age.umap.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/UMAP_ageing_raster.pdf",
       height=5.75, width=16.75, useDingbats=FALSE)

age.umap.p
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
graph1.mean <- unlist(by(data=tec.umap.merge$V1, INDICES=tec.umap.merge$TECtype, FUN=mean))
graph1.order <- names(graph1.mean)
graph2.mean <- unlist(by(data=tec.umap.merge$V2, INDICES=tec.umap.merge$TECtype, FUN=mean))
graph2.order <- names(graph2.mean)

graph.positions <- data.frame("Label"=graph1.order, "V1"=as.numeric(graph1.mean), "V2"=as.numeric(graph2.mean))
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=6.75, fig.width=9.75}
zsg.graph <- ggplot(tec.umap.merge, aes(x=V1, y=V2)) +
  geom_point_rast(aes(color=TECtype), size=2.5) +
  # geom_label_repel(data=graph.positions, 
  #                  aes(label=Label), fill='grey80',
  #                  size=4, colour='black') +
  theme_mike() +
  scale_color_manual(values=inter.cols) +
  theme(strip.background=element_rect(fill='white', colour='white'),
        strip.text=element_text(size=16), 
        axis.title = element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
	      axis.line=element_blank(),
        panel.grid=element_blank(),
        legend.position="right") +
  guides(color=guide_legend(ncol=1,
                            title="TEC.Cluster", title.theme=element_text(size=14),
                            title.position="top",
                            override.aes=list(size=5))) +
  NULL

ggsave(zsg.graph,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ZsG_graph-cluster_interpretation.pdf",
       useDingbats=FALSE, height=6.75, width=9.75)

zsg.graph
```

The labels are too crowded.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Psmb11, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Psmb11 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Psmb11_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Cd52, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Cd52 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cd52_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Aire, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Aire expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Aire_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Ccl21a, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Ccl21a expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Ccl21a_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Gper1, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Gper1 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Gper1_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Krt5, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Krt5 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Krt5_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Cd177, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Cd177 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cd177_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Spink5, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Spink5 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Spink5_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Avil, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Avil expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Avil_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Cdk1, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Cdk1 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cdk1_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Syngr1, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Syngr1 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Syngr1_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Dll4, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Dll4 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Dll4_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Prss16, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Prss16 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Prss16_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Ctsl, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Ctsl expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Ctsl_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Pdpn, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Pdpn expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Pdpn_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.95, fig.width=6.95}
psmb11.p <- ggplot(tec.umap.merge, aes(x=ZsG.Cluster, y=Cxcl12, fill=TECtype)) +
  geom_boxplot(width=0.75, outlier.size=0.5) +
  theme_mike() +
  scale_fill_manual(values=inter.cols) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  guides(fill=FALSE) +
  labs(x="TEC Cluster", y="Cxcl12 expression")

ggsave(psmb11.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/Cxcl12_by_TEC.pdf",
       height=3.95, width=6.95, useDingbats=FALSE)

psmb11.p
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.75, fig.width=15.75}
sf.umap.p <- ggplot(tec.umap.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point_rast(data=tec.umap.merge[,c("UMAP.1", "UMAP.2")], 
                  color='grey80', size=1, alpha=0.5) +
  geom_point_rast(aes(color=SumFactor), size=2.5) +
  theme_mike() +
  scale_color_distiller(palette="Oranges") +
  facet_wrap(~AgeFactor, ncol=4) +
  theme(strip.background=element_rect(fill='white', colour='white'),
        strip.text=element_text(size=21), 
        legend.position="right") +
  guides(color=guide_colorbar(title="Size Factor", ncol=1, title.position="top",
                              title.theme=element_text(size=16),
                              override.aes=list(size=4))) +
  NULL

ggsave(sf.umap.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/UMAP_SizeFactor_raster.pdf",
       height=5.75, width=16.75, useDingbats=FALSE)

sf.umap.p
```




```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.75, fig.width=15.75}
# merge with nDetected genes
zsg.detected <- read.table("~/Dropbox/AgeingExperiment/ZsG_10X_Exp/ZsG_nDetected.tsv",
                           sep="\t", header=TRUE, stringsAsFactors=FALSE)

tec.detec.merge <- merge(tec.umap.merge, zsg.detected, by='Sample')

ndetect.umap.p <- ggplot(tec.detec.merge, aes(x=UMAP.1, y=UMAP.2)) +
  geom_point_rast(data=tec.detec.merge[,c("UMAP.1", "UMAP.2")], 
                  color='grey80', size=1, alpha=0.5) +
  geom_point_rast(aes(color=Feature.Detected), size=2.5) +
  theme_mike() +
  scale_color_viridis() +
  facet_wrap(~AgeFactor, ncol=4) +
  theme(strip.background=element_rect(fill='white', colour='white'),
        strip.text=element_text(size=21), 
        legend.position="right") +
  guides(color=guide_colorbar(title="Features", ncol=1, title.position="top",
                              direction="vertical",
                              title.theme=element_text(size=16),
                              override.aes=list(size=4))) +
  NULL

ggsave(ndetect.umap.p,
       filename="~/Dropbox/AgeingExperiment/ZsG_10X_Exp/UMAP_genesDetected_raster.pdf",
       height=5.75, width=16.75, useDingbats=FALSE)

ndetect.umap.p
```






